<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<!-- Created by GNU Texinfo 5.2, http://www.gnu.org/software/texinfo/ -->
<head>
<title>R FAQ</title>

<meta name="description" content="R FAQ">
<meta name="keywords" content="R FAQ">
<meta name="resource-type" content="document">
<meta name="distribution" content="global">
<meta name="Generator" content="makeinfo">
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<link href="#Top" rel="start" title="Top">
<link href="#SEC_Contents" rel="contents" title="Table of Contents">
<link href="dir.html#Top" rel="up" title="(dir)">
<style type="text/css">
<!--
a.summary-letter {text-decoration: none}
blockquote.smallquotation {font-size: smaller}
div.display {margin-left: 3.2em}
div.example {margin-left: 3.2em}
div.indentedblock {margin-left: 3.2em}
div.lisp {margin-left: 3.2em}
div.smalldisplay {margin-left: 3.2em}
div.smallexample {margin-left: 3.2em}
div.smallindentedblock {margin-left: 3.2em; font-size: smaller}
div.smalllisp {margin-left: 3.2em}
kbd {font-style:oblique}
pre.display {font-family: inherit}
pre.format {font-family: inherit}
pre.menu-comment {font-family: serif}
pre.menu-preformatted {font-family: serif}
pre.smalldisplay {font-family: inherit; font-size: smaller}
pre.smallexample {font-size: smaller}
pre.smallformat {font-family: inherit; font-size: smaller}
pre.smalllisp {font-size: smaller}
span.nocodebreak {white-space:nowrap}
span.nolinebreak {white-space:nowrap}
span.roman {font-family:serif; font-weight:normal}
span.sansserif {font-family:sans-serif; font-weight:normal}
ul.no-bullet {list-style: none}
body {
    margin-left: 5%;
    margin-right: 5%;
}

h1 {
    background: white;
    color: rgb(25%, 25%, 25%);
    font-family: monospace;
    font-size: xx-large;
    text-align: center;
}

h2 {
    background: white;
    color: rgb(40%, 40%, 40%);
    font-family: monospace;
    font-size: x-large;
    text-align: center;
}

h3 {
    background: white;
    color: rgb(40%, 40%, 40%);
    font-family: monospace;
    font-size: large;
}

h4 {
    background: white;
    color: rgb(40%, 40%, 40%);
    font-family: monospace;
}

span.samp {
    font-family: monospace;
}

span.command {
    font-family: monospace;
}

span.option {
    font-family: monospace;
}

span.file {
    font-family: monospace;
}

span.env {
    font-family: monospace;
}

ul.menu {
    list-style-type: decimal;
}

-->
</style>


</head>

<body lang="en" bgcolor="#FFFFFF" text="#000000" link="#0000FF" vlink="#800080" alink="#FF0000">
<h1 class="settitle" align="center">R FAQ</h1>













 <h2>Frequently Asked Questions on R</h2> <h2>Version
3.1.2014-10-26</h2> <address>Kurt Hornik</address> <p><p><hr><p>
<a name="SEC_Contents"></a>
<h2 class="contents-heading">Table of Contents</h2>

<div class="contents">

<ul class="no-bullet">
  <li><a name="toc-Introduction-1" href="#Introduction">1 Introduction</a>
  <ul class="no-bullet">
    <li><a name="toc-Legalese-1" href="#Legalese">1.1 Legalese</a></li>
    <li><a name="toc-Obtaining-this-document-1" href="#Obtaining-this-document">1.2 Obtaining this document</a></li>
    <li><a name="toc-i_0027-e_0021_002cieY_003d1_002f4-i_002ci_00a9ie-e_00b0_00a9e2" href="#g_t_00ec_009d_00b4-_00eb_00ac_00b8_00ec_0084_009c_00eb_00a5_00bc-_00ec_009d_00b8_00ec_009a_00a9_00ed_0095_0098_00eb_008a_0094-_00eb_00b0_00a9_00eb_00b2_0095">1.3 이 문서를 인용하는 방법</a></li>
    <li><a name="toc-Notation-1" href="#Notation">1.4 Notation</a></li>
    <li><a name="toc-Feedback-1" href="#Feedback">1.5 Feedback</a></li>
  </ul></li>
  <li><a name="toc-R-Basics-1" href="#R-Basics">2 R Basics</a>
  <ul class="no-bullet">
    <li><a name="toc-What-is-R_003f-1" href="#What-is-R_003f">2.1 What is R?</a></li>
    <li><a name="toc-What-machines-does-R-run-on_003f-1" href="#What-machines-does-R-run-on_003f">2.2 What machines does R run on?</a></li>
    <li><a name="toc-What-is-the-current-version-of-R_003f-1" href="#What-is-the-current-version-of-R_003f">2.3 What is the current version of R?</a></li>
    <li><a name="toc-How-can-R-be-obtained_003f-1" href="#How-can-R-be-obtained_003f">2.4 How can R be obtained?</a></li>
    <li><a name="toc-How-can-R-be-installed_003f-1" href="#How-can-R-be-installed_003f">2.5 How can R be installed?</a>
    <ul class="no-bullet">
      <li><a name="toc-How-can-R-be-installed-_0028Unix_002dlike_0029-1" href="#How-can-R-be-installed-_0028Unix_002dlike_0029">2.5.1 How can R be installed (Unix-like)</a></li>
      <li><a name="toc-How-can-R-be-installed-_0028Windows_0029-1" href="#How-can-R-be-installed-_0028Windows_0029">2.5.2 How can R be installed (Windows)</a></li>
      <li><a name="toc-How-can-R-be-installed-_0028Mac_0029-1" href="#How-can-R-be-installed-_0028Mac_0029">2.5.3 How can R be installed (Mac)</a></li>
    </ul></li>
    <li><a name="toc-Are-there-Unix_002dlike-binaries-for-R_003f-1" href="#Are-there-Unix_002dlike-binaries-for-R_003f">2.6 Are there Unix-like binaries for R?</a></li>
    <li><a name="toc-What-documentation-exists-for-R_003f-1" href="#What-documentation-exists-for-R_003f">2.7 What documentation exists for R?</a></li>
    <li><a name="toc-Citing-R-1" href="#Citing-R">2.8 Citing R</a></li>
    <li><a name="toc-What-mailing-lists-exist-for-R_003f-1" href="#What-mailing-lists-exist-for-R_003f">2.9 What mailing lists exist for R?</a></li>
    <li><a name="toc-What-is-CRAN_003f-1" href="#What-is-CRAN_003f">2.10 What is <acronym>CRAN</acronym>?</a></li>
    <li><a name="toc-Can-I-use-R-for-commercial-purposes_003f-1" href="#Can-I-use-R-for-commercial-purposes_003f">2.11 Can I use R for commercial purposes?</a></li>
    <li><a name="toc-Why-is-R-named-R_003f-1" href="#Why-is-R-named-R_003f">2.12 Why is R named R?</a></li>
    <li><a name="toc-What-is-the-R-Foundation_003f-1" href="#What-is-the-R-Foundation_003f">2.13 What is the R Foundation?</a></li>
    <li><a name="toc-What-is-R_002dForge_003f-1" href="#What-is-R_002dForge_003f">2.14 What is R-Forge?</a></li>
  </ul></li>
  <li><a name="toc-R-and-S-1" href="#R-and-S">3 R and S</a>
  <ul class="no-bullet">
    <li><a name="toc-What-is-S_003f-1" href="#What-is-S_003f">3.1 What is S?</a></li>
    <li><a name="toc-What-is-S_002dPLUS_003f-1" href="#What-is-S_002dPLUS_003f">3.2 What is <small>S-PLUS</small>?</a></li>
    <li><a name="toc-What-are-the-differences-between-R-and-S_003f-1" href="#What-are-the-differences-between-R-and-S_003f">3.3 What are the differences between R and S?</a>
    <ul class="no-bullet">
      <li><a name="toc-Lexical-scoping-1" href="#Lexical-scoping">3.3.1 Lexical scoping</a></li>
      <li><a name="toc-Models-1" href="#Models">3.3.2 Models</a></li>
      <li><a name="toc-Others-1" href="#Others">3.3.3 Others</a></li>
    </ul></li>
    <li><a name="toc-Is-there-anything-R-can-do-that-S_002dPLUS-cannot_003f-1" href="#Is-there-anything-R-can-do-that-S_002dPLUS-cannot_003f">3.4 Is there anything R can do that <small>S-PLUS</small> cannot?</a></li>
    <li><a name="toc-What-is-R_002dplus_003f-1" href="#What-is-R_002dplus_003f">3.5 What is R-plus?</a></li>
  </ul></li>
  <li><a name="toc-R-Web-Interfaces-1" href="#R-Web-Interfaces">4 R Web Interfaces</a></li>
  <li><a name="toc-R-Add_002dOn-Packages-1" href="#R-Add_002dOn-Packages">5 R Add-On Packages</a>
  <ul class="no-bullet">
    <li><a name="toc-Which-add_002don-packages-exist-for-R_003f-1" href="#Which-add_002don-packages-exist-for-R_003f">5.1 Which add-on packages exist for R?</a>
    <ul class="no-bullet">
      <li><a name="toc-Add_002don-packages-in-R-1" href="#Add_002don-packages-in-R">5.1.1 Add-on packages in R</a></li>
      <li><a name="toc-Add_002don-packages-from-CRAN-1" href="#Add_002don-packages-from-CRAN">5.1.2 Add-on packages from <acronym>CRAN</acronym></a></li>
      <li><a name="toc-Add_002don-packages-from-Omegahat-1" href="#Add_002don-packages-from-Omegahat">5.1.3 Add-on packages from Omegahat</a></li>
      <li><a name="toc-Add_002don-packages-from-Bioconductor-1" href="#Add_002don-packages-from-Bioconductor">5.1.4 Add-on packages from Bioconductor</a></li>
      <li><a name="toc-Other-add_002don-packages-1" href="#Other-add_002don-packages">5.1.5 Other add-on packages</a></li>
    </ul></li>
    <li><a name="toc-How-can-add_002don-packages-be-installed_003f-1" href="#How-can-add_002don-packages-be-installed_003f">5.2 How can add-on packages be installed?</a></li>
    <li><a name="toc-How-can-add_002don-packages-be-used_003f-1" href="#How-can-add_002don-packages-be-used_003f">5.3 How can add-on packages be used?</a></li>
    <li><a name="toc-How-can-add_002don-packages-be-removed_003f-1" href="#How-can-add_002don-packages-be-removed_003f">5.4 How can add-on packages be removed?</a></li>
    <li><a name="toc-How-can-I-create-an-R-package_003f-1" href="#How-can-I-create-an-R-package_003f">5.5 How can I create an R package?</a></li>
    <li><a name="toc-How-can-I-contribute-to-R_003f-1" href="#How-can-I-contribute-to-R_003f">5.6 How can I contribute to R?</a></li>
  </ul></li>
  <li><a name="toc-R-and-Emacs-1" href="#R-and-Emacs">6 R and Emacs</a>
  <ul class="no-bullet">
    <li><a name="toc-Is-there-Emacs-support-for-R_003f-1" href="#Is-there-Emacs-support-for-R_003f">6.1 Is there Emacs support for R?</a></li>
    <li><a name="toc-Should-I-run-R-from-within-Emacs_003f-1" href="#Should-I-run-R-from-within-Emacs_003f">6.2 Should I run R from within Emacs?</a></li>
    <li><a name="toc-Debugging-R-from-within-Emacs-1" href="#Debugging-R-from-within-Emacs">6.3 Debugging R from within Emacs</a></li>
  </ul></li>
  <li><a name="toc-R-Miscellanea-1" href="#R-Miscellanea">7 R Miscellanea</a>
  <ul class="no-bullet">
    <li><a name="toc-How-can-I-set-components-of-a-list-to-NULL_003f-1" href="#How-can-I-set-components-of-a-list-to-NULL_003f">7.1 How can I set components of a list to NULL?</a></li>
    <li><a name="toc-How-can-I-save-my-workspace_003f-1" href="#How-can-I-save-my-workspace_003f">7.2 How can I save my workspace?</a></li>
    <li><a name="toc-How-can-I-clean-up-my-workspace_003f-1" href="#How-can-I-clean-up-my-workspace_003f">7.3 How can I clean up my workspace?</a></li>
    <li><a name="toc-How-can-I-get-eval_0028_0029-and-D_0028_0029-to-work_003f-1" href="#How-can-I-get-eval_0028_0029-and-D_0028_0029-to-work_003f">7.4 How can I get eval() and D() to work?</a></li>
    <li><a name="toc-Why-do-my-matrices-lose-dimensions_003f-1" href="#Why-do-my-matrices-lose-dimensions_003f">7.5 Why do my matrices lose dimensions?</a></li>
    <li><a name="toc-How-does-autoloading-work_003f-1" href="#How-does-autoloading-work_003f">7.6 How does autoloading work?</a></li>
    <li><a name="toc-How-should-I-set-options_003f-1" href="#How-should-I-set-options_003f">7.7 How should I set options?</a></li>
    <li><a name="toc-How-do-file-names-work-in-Windows_003f-1" href="#How-do-file-names-work-in-Windows_003f">7.8 How do file names work in Windows?</a></li>
    <li><a name="toc-Why-does-plotting-give-a-color-allocation-error_003f-1" href="#Why-does-plotting-give-a-color-allocation-error_003f">7.9 Why does plotting give a color allocation error?</a></li>
    <li><a name="toc-How-do-I-convert-factors-to-numeric_003f-1" href="#How-do-I-convert-factors-to-numeric_003f">7.10 How do I convert factors to numeric?</a></li>
    <li><a name="toc-Are-Trellis-displays-implemented-in-R_003f-1" href="#Are-Trellis-displays-implemented-in-R_003f">7.11 Are Trellis displays implemented in R?</a></li>
    <li><a name="toc-What-are-the-enclosing-and-parent-environments_003f-1" href="#What-are-the-enclosing-and-parent-environments_003f">7.12 What are the enclosing and parent environments?</a></li>
    <li><a name="toc-How-can-I-substitute-into-a-plot-label_003f-1" href="#How-can-I-substitute-into-a-plot-label_003f">7.13 How can I substitute into a plot label?</a></li>
    <li><a name="toc-What-are-valid-names_003f-1" href="#What-are-valid-names_003f">7.14 What are valid names?</a></li>
    <li><a name="toc-Are-GAMs-implemented-in-R_003f-1" href="#Are-GAMs-implemented-in-R_003f">7.15 Are GAMs implemented in R?</a></li>
    <li><a name="toc-Why-is-the-output-not-printed-when-I-source_0028_0029-a-file_003f-1" href="#Why-is-the-output-not-printed-when-I-source_0028_0029-a-file_003f">7.16 Why is the output not printed when I source() a file?</a></li>
    <li><a name="toc-Why-does-outer_0028_0029-behave-strangely-with-my-function_003f-1" href="#Why-does-outer_0028_0029-behave-strangely-with-my-function_003f">7.17 Why does outer() behave strangely with my function?</a></li>
    <li><a name="toc-Why-does-the-output-from-anova_0028_0029-depend-on-the-order-of-factors-in-the-model_003f-1" href="#Why-does-the-output-from-anova_0028_0029-depend-on-the-order-of-factors-in-the-model_003f">7.18 Why does the output from anova() depend on the order of factors in the model?</a></li>
    <li><a name="toc-How-do-I-produce-PNG-graphics-in-batch-mode_003f-1" href="#How-do-I-produce-PNG-graphics-in-batch-mode_003f">7.19 How do I produce PNG graphics in batch mode?</a></li>
    <li><a name="toc-How-can-I-get-command-line-editing-to-work_003f-1" href="#How-can-I-get-command-line-editing-to-work_003f">7.20 How can I get command line editing to work?</a></li>
    <li><a name="toc-How-can-I-turn-a-string-into-a-variable_003f-1" href="#How-can-I-turn-a-string-into-a-variable_003f">7.21 How can I turn a string into a variable?</a></li>
    <li><a name="toc-Why-do-lattice_002ftrellis-graphics-not-work_003f-1" href="#Why-do-lattice_002ftrellis-graphics-not-work_003f">7.22 Why do lattice/trellis graphics not work?</a></li>
    <li><a name="toc-How-can-I-sort-the-rows-of-a-data-frame_003f-1" href="#How-can-I-sort-the-rows-of-a-data-frame_003f">7.23 How can I sort the rows of a data frame?</a></li>
    <li><a name="toc-Why-does-the-help_002estart_0028_0029-search-engine-not-work_003f-1" href="#Why-does-the-help_002estart_0028_0029-search-engine-not-work_003f">7.24 Why does the help.start() search engine not work?</a></li>
    <li><a name="toc-Why-did-my-_002eRprofile-stop-working-when-I-updated-R_003f-1" href="#Why-did-my-_002eRprofile-stop-working-when-I-updated-R_003f">7.25 Why did my .Rprofile stop working when I updated R?</a></li>
    <li><a name="toc-Where-have-all-the-methods-gone_003f-1" href="#Where-have-all-the-methods-gone_003f">7.26 Where have all the methods gone?</a></li>
    <li><a name="toc-How-can-I-create-rotated-axis-labels_003f-1" href="#How-can-I-create-rotated-axis-labels_003f">7.27 How can I create rotated axis labels?</a></li>
    <li><a name="toc-Why-is-read_002etable_0028_0029-so-inefficient_003f-1" href="#Why-is-read_002etable_0028_0029-so-inefficient_003f">7.28 Why is read.table() so inefficient?</a></li>
    <li><a name="toc-What-is-the-difference-between-package-and-library_003f-1" href="#What-is-the-difference-between-package-and-library_003f">7.29 What is the difference between package and library?</a></li>
    <li><a name="toc-I-installed-a-package-but-the-functions-are-not-there-1" href="#I-installed-a-package-but-the-functions-are-not-there">7.30 I installed a package but the functions are not there</a></li>
    <li><a name="toc-Why-doesn_0027t-R-think-these-numbers-are-equal_003f-1" href="#Why-doesn_0027t-R-think-these-numbers-are-equal_003f">7.31 Why doesn&rsquo;t R think these numbers are equal?</a></li>
    <li><a name="toc-How-can-I-capture-or-ignore-errors-in-a-long-simulation_003f-1" href="#How-can-I-capture-or-ignore-errors-in-a-long-simulation_003f">7.32 How can I capture or ignore errors in a long simulation?</a></li>
    <li><a name="toc-Why-are-powers-of-negative-numbers-wrong_003f-1" href="#Why-are-powers-of-negative-numbers-wrong_003f">7.33 Why are powers of negative numbers wrong?</a></li>
    <li><a name="toc-How-can-I-save-the-result-of-each-iteration-in-a-loop-into-a-separate-file_003f-1" href="#How-can-I-save-the-result-of-each-iteration-in-a-loop-into-a-separate-file_003f">7.34 How can I save the result of each iteration in a loop into a separate file?</a></li>
    <li><a name="toc-Why-are-p_002dvalues-not-displayed-when-using-lmer_0028_0029_003f-1" href="#Why-are-p_002dvalues-not-displayed-when-using-lmer_0028_0029_003f">7.35 Why are <em>p</em>-values not displayed when using lmer()?</a></li>
    <li><a name="toc-Why-are-there-unwanted-borders_002c-lines-or-grid_002dlike-artifacts-when-viewing-a-plot-saved-to-a-PS-or-PDF-file_003f" href="#Why-are-there-unwanted-borders">7.36 Why are there unwanted borders, lines or grid-like artifacts when viewing a plot saved to a PS or PDF file?</a></li>
    <li><a name="toc-Why-does-backslash-behave-strangely-inside-strings_003f-1" href="#Why-does-backslash-behave-strangely-inside-strings_003f">7.37 Why does backslash behave strangely inside strings?</a></li>
    <li><a name="toc-How-can-I-put-error-bars-or-confidence-bands-on-my-plot_003f-1" href="#How-can-I-put-error-bars-or-confidence-bands-on-my-plot_003f">7.38 How can I put error bars or confidence bands on my plot?</a></li>
    <li><a name="toc-How-do-I-create-a-plot-with-two-y_002daxes_003f-1" href="#How-do-I-create-a-plot-with-two-y_002daxes_003f">7.39 How do I create a plot with two y-axes?</a></li>
    <li><a name="toc-How-do-I-access-the-source-code-for-a-function_003f-1" href="#How-do-I-access-the-source-code-for-a-function_003f">7.40 How do I access the source code for a function?</a></li>
    <li><a name="toc-Why-does-summary_0028_0029-report-strange-results-for-the-R_005e2-estimate-when-I-fit-a-linear-model-with-no-intercept_003f-1" href="#Why-does-summary_0028_0029-report-strange-results-for-the-R_005e2-estimate-when-I-fit-a-linear-model-with-no-intercept_003f">7.41 Why does summary() report strange results for the R^2 estimate when I fit a linear model with no intercept?</a></li>
    <li><a name="toc-Why-is-R-apparently-not-releasing-memory_003f-1" href="#Why-is-R-apparently-not-releasing-memory_003f">7.42 Why is R apparently not releasing memory?</a></li>
  </ul></li>
  <li><a name="toc-R-Programming-1" href="#R-Programming">8 R Programming</a>
  <ul class="no-bullet">
    <li><a name="toc-How-should-I-write-summary-methods_003f-1" href="#How-should-I-write-summary-methods_003f">8.1 How should I write summary methods?</a></li>
    <li><a name="toc-How-can-I-debug-dynamically-loaded-code_003f-1" href="#How-can-I-debug-dynamically-loaded-code_003f">8.2 How can I debug dynamically loaded code?</a></li>
    <li><a name="toc-How-can-I-inspect-R-objects-when-debugging_003f-1" href="#How-can-I-inspect-R-objects-when-debugging_003f">8.3 How can I inspect R objects when debugging?</a></li>
    <li><a name="toc-How-can-I-change-compilation-flags_003f-1" href="#How-can-I-change-compilation-flags_003f">8.4 How can I change compilation flags?</a></li>
    <li><a name="toc-How-can-I-debug-S4-methods_003f-1" href="#How-can-I-debug-S4-methods_003f">8.5 How can I debug S4 methods?</a></li>
  </ul></li>
  <li><a name="toc-R-Bugs-1" href="#R-Bugs">9 R Bugs</a>
  <ul class="no-bullet">
    <li><a name="toc-What-is-a-bug_003f-1" href="#What-is-a-bug_003f">9.1 What is a bug?</a></li>
    <li><a name="toc-How-to-report-a-bug-1" href="#How-to-report-a-bug">9.2 How to report a bug</a></li>
  </ul></li>
  <li><a name="toc-Acknowledgments-1" href="#Acknowledgments">10 Acknowledgments</a></li>
</ul>
</div>






<p><strong>NOTE</strong> from the maintainer Chel Hee Lee &lt;<a href="mailto:chl948@mail.usask.ca">chl948@mail.usask.ca</a>&gt;
</p>
<p>This translation is updated on a day-by-day basis (Thu Dec 25 09:36:45 CST 2014).  The initial work was done with R-2.8.1 (21-DEC-2008) by Chel Hee Lee.  The latest version is based on the Revision: 67221 (3.2.0 Under development (unstable)). 
</p>
<p>This work is a result of cumulative efforts made by the contributors listed in the below:
</p><ul>
<li> Chel Hee Lee &lt;<a href="mailto:chl948@mail.usask.ca">chl948@mail.usask.ca</a>&gt;, University of Saskatchewan, Saskatoon, Saskatchewan, Canada, 2008&ndash;2014
</li><li> Heather Kim &lt;<a href="mailto:heatherkimca@gmail.com">heatherkimca@gmail.com</a>&gt;, University of Mantioba, Winnipeg, Manitoba, Canada, 2013
</li></ul>

<p>If you wish to voluntarily participate in this translation project, please visit the page at <a href="http://r-forge.r-project.org/projects/ihelp/">http://r-forge.r-project.org/projects/ihelp/</a>
</p>



<a name="Top"></a>
<a name="R-FAQ"></a>
<h1 class="top">R FAQ</h1>


<hr>
<a name="Introduction"></a>
<a name="Introduction-1"></a>
<h2 class="chapter">1 Introduction</h2>

<p>이 문서는 R에 관하여 가장 많이 자주 물어보는 질문들에 대해서 정리한 문서입니다.
</p>

<hr>
<a name="Legalese"></a>
<a name="Legalese-1"></a>
<h3 class="section">1.1 Legalese</h3>

<p>이 문서의 영문본에 대한 저작권은 &copy; 1998&ndash;2014 Kurt Hornik에게 있으며,
한국어 번역본은 Chel Hee Lee &copy; 2009&ndash;2014에 있습니다.
</p>
<p>본 문서는 자유소프트웨어(free software)이므로, 자유소프트웨어 재단 (Free Software Foundation)에 의해서
배포되는 <acronym>GNU</acronym> 일반공중라이센스 (General Public License) 버전 2 또는 그 이후의 버전들을 통하여
수정 및 재배포를 할 수 있습니다. 
</p>
<p>이 문서가 사용자에게 유용하길 바라는 희망에서 배포되고 있습니다. 그러나, 어떤 상업적 형태(MERCHANTABILITY)로서의 보증 및
특수한 목적에 적합(FITNESS FOR A PARTICULAR PURPOSES)등 어떠한 형태의 보증을 하지 않습니다.  보다 자세한
사항들은 <acronym>GNU</acronym> 일반공증라이센스를 참조하시길 부탁드립니다. 
</p>
<p><acronym>GNU</acronym> 일반공증라이센스의 사본들을 아래의 주소로부터 찾으실 수 있습니다.
</p>
<div class="display">
<pre class="display"><a href="http://www.R-project.org/Licenses/">http://www.R-project.org/Licenses/</a>
</pre></div>


<hr>
<a name="Obtaining-this-document"></a>
<a name="Obtaining-this-document-1"></a>
<h3 class="section">1.2 Obtaining this document</h3>

<p>가장 최신 버전의 문서는 항상 아래의 주소에서 확인이 가능합니다.
</p>
<div class="display">
<pre class="display"><a href="http://CRAN.R-project.org/doc/FAQ/">http://CRAN.R-project.org/doc/FAQ/</a>
</pre></div>

<p><a href="http://CRAN.R-project.org/doc/FAQ/R-FAQ.txt">plain <acronym>ASCII</acronym>
text</a>, <a href="http://CRAN.R-project.org/doc/FAQ/R-FAQ.dvi.gz">DVI</a>,
<a href="http://CRAN.R-project.org/doc/FAQ/R-FAQ.info.gz"><acronym>GNU</acronym> info</a>,
<a href="http://CRAN.R-project.org/doc/FAQ/R-FAQ.html"><acronym>HTML</acronym></a>,
<a href="http://CRAN.R-project.org/doc/FAQ/R-FAQ.pdf">PDF</a>,
<a href="http://CRAN.R-project.org/doc/FAQ/R-FAQ.ps.gz">PostScript</a> 으로 변환되어진
문서형식을 가진 버전들을 얻을 수 있습니다.  또한, 이렇게 다양한 형식의 문서들을 생성하기 위해 사용된
<a href="http://CRAN.R-project.org/doc/FAQ/R-FAQ.texi">Texinfo source</a> 를 얻을 수
있습니다.
</p>
<p>그외에도 R <acronym>FAQ</acronym> 문서를 <acronym>CRAN</acronym> 사이트 (see <a href="#What-is-CRAN_003f">What is CRAN?</a> 참조)의 하위디렉토리
<samp>doc/FAQ</samp>에서 찾을 수 있습니다.
</p>
<hr>
<a name="g_t_00ec_009d_00b4-_00eb_00ac_00b8_00ec_0084_009c_00eb_00a5_00bc-_00ec_009d_00b8_00ec_009a_00a9_00ed_0095_0098_00eb_008a_0094-_00eb_00b0_00a9_00eb_00b2_0095"></a>
<a name="i_0027-e_0021_002cieY_003d1_002f4-i_002ci_00a9ie-e_00b0_00a9e2"></a>
<h3 class="section">1.3 이 문서를 인용하는 방법</h3>

<p>퍼블리케이션 (publication)에서 본 <acronym>FAQ</acronym> 문서를 인용하고자 할 때에는 Hornik
(2014), &ldquo;The R <acronym>FAQ</acronym>&rdquo; 이라고 해주시고, 위에서 언급한 <em>공식적인</em>
<acronym>URL</acronym>과 ISBN {No value for &lsquo;FAQ_ISBN&rsquo;}을 기재해 주세요.
</p>
<div class="example">
<pre class="example">@Misc{,
  author        = {Kurt Hornik},
  title         = {{R} {FAQ}},
  year          = {2014},
  url           = {http://CRAN.R-project.org/doc/FAQ/R-FAQ.html}
}
</pre></div>


<hr>
<a name="Notation"></a>
<a name="Notation-1"></a>
<h3 class="section">1.4 Notation</h3>

<p>이 문서에서는 거의 대부분이 표준에 가까운 표기법을 사용합니다.  예를들면, &lsquo;<samp>R&gt;</samp>&rsquo;은 R 프롬프트(prompt)를 나타내고,
&lsquo;<samp>$</samp>&rsquo;는 쉘프롬프트(shell prompt)를 의미합니다.
</p>
<hr>
<a name="Feedback"></a>
<a name="Feedback-1"></a>
<h3 class="section">1.5 Feedback</h3>

<p>본 문서에 대한 피드백은 <a href="mailto:gnustats@korea.gnu.org">gnustats@korea.gnu.org</a> 또는
<a href="mailto:ihelp-translators@lists.r-forge.r-project.org">ihelp-translators@lists.r-forge.r-project.org</a>로 보내주시길 부탁드립니다.
</p>
<p>특히, 다음의 내용에 대해서 알아주셨으면 합니다.  이 문서의 작성자 (및 번역자 Chel Hee Lee)는 Windows 또는
Macintosh 시스템들을 가지고 있지 않습니다.  따라서, Windows와 Mac OS X와 관련된 특별한 내용들에 대해서는
<a href="http://CRAN.R-project.org/bin/windows/base/rw-FAQ.html">&ldquo;R for Windows
<acronym>FAQ</acronym>&rdquo;</a> 와
<a href="http://CRAN.R-project.org/bin/macosx/RMacOSX-FAQ.html">&ldquo;R for Mac OS X
<acronym>FAQ</acronym>&rdquo;</a>을 참고 해주세요.  만약 Macintosh와 Windows 시스템들에 관련된 정보라고 할지라도, 이
문서내에 곡 포함되어야 한다고 생각되시면 꼭 알려주시기를 부탁드립니다.
</p>

<hr>
<a name="R-Basics"></a>
<a name="R-Basics-1"></a>
<h2 class="chapter">2 R Basics</h2>


<hr>
<a name="What-is-R_003f"></a>
<a name="What-is-R_003f-1"></a>
<h3 class="section">2.1 What is R?</h3>

<p>R은 통계 계산 (statistical computation)과 그래픽 (graphics)를 위한 시스템입니다.  이것은
언어(language), 그래픽을 사용하는 실행환경 (run-time environment), 디버거(debugger), 특정시스템
함수들에 대한 접근, 그리고 스크립트 파일들에 저장되어 있는 프로그램들을 실행시키는 요소들로 구성되어 있습니다.
</p>
<p>R의 디자인은 Becker, Chambers &amp; Wilks의 S (see <a href="#What-is-S_003f">What is S?</a>)와 Sussman의
<a href="http://www.cs.indiana.edu/scheme-repository/home.html">Scheme</a>으로부터 크게
영향을 받았습니다. 결과적으로 보이는 언어는 S의 외형과 매우 흡사하지만, 근본적인 구현방법과 시맨틱스(sematics)는
Scheme으로부터 유래되었습니다. 더 많은 내용을 알고 싶으시다면 <a href="#What-are-the-differences-between-R-and-S_003f">What are the differences between R and S?</a>를 참고하시길 바랍니다.
</p>
<p>R의 핵심은 인터프리트된 컴퓨터 언어 (interpreted computer language)입니다.  이것은 분기(branching),
루핑(Looping) 그리고 함수(functions)를 이용한 모듈(modular) 프로그래밍을 가능하게 합니다.  R에서 사용자가
사용하는 함수들의 대부분은 R로 작성되었습니다.  또한, 사용자가 효율성(efficiency)를 위한 목적으로 C, C++, 또는
FORTRAN과 같은 언어로 작성된 프로시져들에 인터페이싱(interface)하는 것이 가능합니다.  R 배포판
(distribution)은 방대한 양의 통계적 프로시져(statistical procedure)들을 위한
기능(functionalities)들을 포함하고 있습니다.  이들중에는 선형 및 일반화 선형 모델 (linear and
generalized linear models), 비선형모델 (nonlinear regression model), 시계열분석 (time
series analysis), 전통적인 모수 및 비모수 테스트 (classical parametric and nonparametric
tests), 클러스터링 (clustering)과 평활화 (smoothing)이 있습니다.  또한 매우 다양한 종류의 데이터 시각적
표현들을 생성할 수 있는 매우 유연한 그래픽 환경을 제공해주는 방대한 양의 함수들의 집합이 있습니다.  특수한 목적을 달성하기 위해서
개발된 애드온 패키지(&ldquo;add-on packages&rdquo;)로 알려져 있는 부가적인 모듈들 역시 이용이 가능합니다.  (이들에 대해서 더
알고 싶으시면 <a href="#R-Add_002dOn-Packages">R Add-On Packages</a>를 참고해주세요).
</p>
<p>최초의 R은 <a href="mailto:Ross.Ihaka@R-project.org">Ross Ihaka</a>와
<a href="mailto:Robert.Gentleman@R-project.org">Robert Gentleman</a>가 New Zealand의
Auckland 에 있는 University of Auckland의 통계학과에 재직중일 당시 작성되었습니다.  그 이후, 많은 사용자들이
코드를 제공하고 버그를 리포팅 함으로서 R에 기여 해왔습니다. 
</p>
<p>1997년 중반 이래로 &ldquo;R Core Team&rdquo;이라는 이름의 그룹에 의해서 R의 소스 코드 아카이브를 수정 및 관리되어지고
있습니다.  이 그룹은 Doug Bates, John Chambers, Peter Dalgaard, Seth Falcon, Robert
Gentleman, Kurt Hornik, Stefano Iacus, Ross Ihaka, Friedrich Leisch, Uwe
Ligges, Thomas Lumley, Martin Maechler, Duncan Murdoch, Paul Murrell, Martyn
Plummer, Brian Ripley, Deepayan Sarkar, Duncan Temple Lang, Luke Tierney,
Simon Urbanek 으로 구성되어 있습니다. 
</p>
<p>R은 홈페이지 <a href="http://www.R-project.org/">http://www.R-project.org/</a>를 가지고 있습니다.  <acronym>GNU</acronym>
<a href="http://www.gnu.org/copyleft/copyleft.html">copyleft</a>를 기반으로 배포되는
<a href="http://www.gnu.org/philosophy/free-sw.html">free software</a>입니다.  이는
&ldquo;<acronym>GNU</acronym> S&rdquo;라고 불리는 <a href="http://www.gnu.org/"><acronym>GNU</acronym></a> 공식 프로젝트의
일부입니다.
</p>
<hr>
<a name="What-machines-does-R-run-on_003f"></a>
<a name="What-machines-does-R-run-on_003f-1"></a>
<h3 class="section">2.2 What machines does R run on?</h3>

<p>R은 Unix, Windows와 Mac 계열의 환경에서 개발되어져 왔습니다.  R 1.7.1 이후로는 Mac OS Classic 에 대한
지원이 종료되었습니다.
</p>
<p>현 버전의 R은 보편적인 유닉스와 같은 (<a href="http://en.wikipedia.org/wiki/Unix-like">http://en.wikipedia.org/wiki/Unix-like</a>) 플랫폼을
기반으로 개발되고 설정됩니다. <var>cpu</var>-linux-gnu for the i386, amd64, alpha, arm/armel,
hppa, ia64, m68k, mips/mipsel, powerpc, s390 and sparc <acronym>CPU</acronym>s (e.g.,
<a href="http://buildd.debian.org/build.php?&amp;pkg=r-base">http://buildd.debian.org/build.php?&amp;pkg=r-base</a>), i386-hurd-gnu,
<var>cpu</var>-kfreebsd-gnu for i386 and amd64, powerpc-apple-darwin,
mips-sgi-irix, rs6000-ibm-aix, 그리고 sparc-sun-solaris이 이에 해당합니다. 
</p>

<p>If you know about other platforms, please drop us a note.
</p>
<hr>
<a name="What-is-the-current-version-of-R_003f"></a>
<a name="What-is-the-current-version-of-R_003f-1"></a>
<h3 class="section">2.3 What is the current version of R?</h3>
 
<p>현재 발표된 (released) 버전은 3.1.2입니다.  &lsquo;major.minor.patchlevel&rsquo;의 형식을
가지는 번호체계(numbering scheme)에 기초하여, R은 두 개발버전 (development version)들과 현재 발표된
버전의 일부를 수정한 &rsquo;r-patched&rsquo;라는 이름의 패치된 버전이 있으며, 다음 단계의 마이너(minor) 버전 또는 결과적으로는
주버전(major)이 될 &lsquo;r-devel&rsquo;라는 이름으로 발표된 R이 있습니다.  패치 버전 (r-patched)은 대개 버그 수정을 한
것입니다.  새로운 기능 및 형상들은 일반적으로 r-devel 에서 소개됩니다. 
</p>
<hr>
<a name="How-can-R-be-obtained_003f"></a>
<a name="How-can-R-be-obtained_003f-1"></a>
<h3 class="section">2.4 How can R be obtained?</h3>

<p>R 원시코드 (sources), 바이너리 (binaries), 그리고 문서들(documentation)은
<acronym>CRAN</acronym>(&ldquo;Comprehensive R Archive Network&rdquo;)을 통하여 얻을 수 있습니다. (<a href="#What-is-CRAN_003f">What is CRAN?</a>을 살펴보세요).
</p>
<p>원시코드 (sources)는 또한 R 서브버전 저장소 (subversion repository)인
<a href="https://svn.R-project.org/R/">https://svn.R-project.org/R/</a>를 통하여 얻을 수 있습니다.  그러나, 익명 rsync (그리고 CVS)를
사용할 수 없습니다.
</p>
<p>R의 r-devel와 r-patched 버전들에 대한 일일 스냅샷(daily snapshot)을
<a href="ftp://ftp.stat.math.ethz.ch/Software/R">ftp://ftp.stat.math.ethz.ch/Software/R</a>에서 타르볼(tarball)의 형태로 찾을 수 있습니다. 
</p>





<hr>
<a name="How-can-R-be-installed_003f"></a>
<a name="How-can-R-be-installed_003f-1"></a>
<h3 class="section">2.5 How can R be installed?</h3>


<hr>
<a name="How-can-R-be-installed-_0028Unix_002dlike_0029"></a>
<a name="How-can-R-be-installed-_0028Unix_002dlike_0029-1"></a>
<h4 class="subsection">2.5.1 How can R be installed (Unix-like)</h4>

<p>만약 R이 이미 설치되어 있다면, 쉘 프롬프트(shell prompt)에서 <kbd>R</kbd>이라고 입력을 함으로서 실행할 수 있습니다
(물론, 실행가능한 파일이 경로에 올바르게 위치해 있다는 가정하에서 입니다). 
</p>
<p>만약 사용자의 플랫폼에 바이너리(binaries)가 사용가능하다면 (<a href="#Are-there-Unix_002dlike-binaries-for-R_003f">Are there Unix-like binaries for R?</a> 섹션을 살펴보세요), 이와 함께 제공되는 지시사항들을 따른다면 이들을 사용할 수 있습니다.
</p>
<p>그렇지 않다면, R을 직접 컴파일하여 설치하여야 합니다.  일반적인 Unix와 같은 플랫폼(<a href="#What-machines-does-R-run-on_003f">What machines does R run on?</a>를 참고해주세요)에서 이 과정을 수행하는 것은 매우 쉽쉽니다.  R의 배포판과 함께 제공되는 <samp>INSTALL</samp>
파일은 간단한 설치지침을 포함하고 있으면 &ldquo;R Installation and Administration&rdquo;라는 가이드는 설치와 관련한
자세한 내용들이 잘 문서화 되어 있습니다 (<a href="#What-documentation-exists-for-R_003f">What documentation exists for R?</a>을 참고하세요).
</p>
<p>참고로 R을 빌드(build)하기 위해서는 C 컴파일러(compiler) 외에도 FORTRAN 컴파일러 혹은 아마도
<code>f2c</code>라는 것이 필요할 수도 있습니다.
</p>
<p>가장 간단한 방법은 R 원시코드(source code)의 압축을 풀고, 압축이 풀어 생성된 디렉토리로 이동한 후, (쉘 프롬프트상에서)
다음과 같은 명령어를 입력하는 것입니다.
</p>
<div class="example">
<pre class="example">$ ./configure
$ make
</pre></div>

<p>만약 명령어들이 성공적으로 수행된다면, R 바이너리와 쉘 스크립트로 작성된 사용자가 이용이 가능한 <samp>R</samp>이라는 이름을 가진
파일이 생성되고, 이는 <samp>bin</samp> 디렉토리에 복사되어 질 것입니다.  사용자는 이 스크립트를 실행시키고자 하는 위치로 스크립트를
복사해 넣기만 하면 됩니다.  예를들면, <samp>/usr/local/bin</samp> 디렉토리입니다.  또한, 텍스트로만 구성된 도움말 페이지
뿐만아니라 <acronym>HTML</acronym>, LaTeX 버전의 문서들도 함께 설치됩니다. 
</p>
<p><samp>doc/manual</samp>에 놓여있는 <samp>refman.dvi</samp> (R object reference index) 혹은
<samp>R-exts.dvi</samp> (&ldquo;R Extension Writers Guide&rdquo;)와 같은 R 매뉴얼들의 DVI 버전을 생성하기
위해서는 <kbd>make dvi</kbd>를 이용해보세요.  이러한 파일들은 <code>xdvi</code> 또는 <code>dvips</code>와 같은
표준 프로그램들을 이용하여 미리 보거나 프린트를 할 수 있습니다.  또한, <kbd>make pdf</kbd>의 사용은 매뉴얼들을 PDF
(Portable Document Format)의 형식으로 생성하게 해줍니다.  이들은 Arobat을 이용하여 볼 수 있습니다.
<acronym>GNU</acronym> Texinfo 시스템에 의하여 작성된 매뉴얼들 (본 한국어 매뉴얼 역시 <acronym>GNU</acronym> Texinfo
시스템을 이용하여 작성되었습니다) 은 또한 Emacs 또는 독립적인 <acronym>GNU</acronym> Info 와 함께 온라인으로 읽을 수 읽을
수 있도록 info 파일들로 변환할 수 있습니다.  이들을 생성하기 위해서는 <kbd>make info</kbd>를 이용하시길 바랍니다. (단,
이경우에는 Makeinfo 의 버전이 4.5 이상이 요구됩니다).
</p>
<p>마지막으로 R 시스템이 정상적으로 작동하는지 확인하기 위해서 <kbd>make check</kbd>를 이용해 보세요.
</p>
<p><kbd>make install</kbd>을 이용하여 &ldquo;system-wide&rdquo;(시스템 전체에서 사용이 가능한) 설치를 수행할 수도
있습니다. 기본적으로 이것은 다음의 디렉토리에 설치됩니다. 
</p>
<dl compact="compact">
<dt><samp>${prefix}/bin</samp></dt>
<dd><p>사용자가 직접 이용하는 쉘 스크립트
</p></dd>
<dt><samp>${prefix}/man/man1</samp></dt>
<dd><p>the man page
</p></dd>
<dt><samp>${prefix}/lib/R</samp></dt>
<dd><p>위의 내용을 제외한 나머지 모두 (라이브러리, 온라인 도움말, &hellip;)이 이곳에 설치됩니다. 이것을 설치된 시스템의 &ldquo;R
Home Directory&rdquo;(R 홈디렉토리, <code>R_HOME</code>)이라고 합니다. 
</p></dd>
</dl>

<p>위에서 <code>prefix</code>는 환경설정을 하는 동안에 결정되어 지며 (일반적으로 <samp>/usr/local/</samp>입니다), 아래에
보이는 것과 같이 <code>configure</code>에 옵션을 이용하여 설치시에 정할 수도 있습니다.
</p>
<div class="example">
<pre class="example">$ ./configure --prefix=/where/you/want/R/to/go
</pre></div>

<p>(즉, R의 실행파일들은 <samp>/where/you/want/R/to/go/bin</samp>에 설치될 것입니다).
</p>
<p>DVI, info, PDF 버전의 매뉴얼들을 설치하기 위해서는 <kbd>make install-dvi</kbd>, <kbd>make
install-info</kbd>, <kbd>make install-pdf</kbd>을 각각 실행시키세요.
</p>
<hr>
<a name="How-can-R-be-installed-_0028Windows_0029"></a>
<a name="How-can-R-be-installed-_0028Windows_0029-1"></a>
<h4 class="subsection">2.5.2 How can R be installed (Windows)</h4>

<p><acronym>CRAN</acronym> 사이트의 <samp>bin/windows</samp> 디렉토리는 ix86 과 x86_64 칩상에서 돌아가는 Windows 2000 과
그 이후 의 버전들 (64비트 포함)에서 돌아갈 수 있는 R base 배포판과 방대한 양의 애드온(add-on) 패키지들에 대한
바이너리들을 포함하고 있습니다.  Windows 버전의 R은 Robert Gentleman과 Guido Masarotto에 의해서
만들어졌으며, 현재는 <a href="mailto:murdoch@stats.uwo.ca">Duncan Murdoch</a>와
<a href="mailto:Brian.Ripley@R-project.org">Brian D. Ripley</a>에 의해서 개발 및 유지보수가 되고
있습니다.
</p>
<p>대부분의 경우, Windows 설치 프로그램 (installer)은 이용하기에 가장 쉬운 도구일 것입니다.
</p>
<p>보다 더 자세한 사항에 대해서는
<a href="http://CRAN.R-project.org/bin/windows/base/rw-FAQ.html">&ldquo;R for
Windows <acronym>FAQ</acronym>&rdquo;</a>를 참고하세요.
</p>
<hr>
<a name="How-can-R-be-installed-_0028Mac_0029"></a>
<a name="How-can-R-be-installed-_0028Mac_0029-1"></a>
<h4 class="subsection">2.5.3 How can R be installed (Mac)</h4>

<p><acronym>CRAN</acronym> 사이트내의 <samp>bin/macosx</samp> 디렉토리는 <samp>R.dmg</samp>라는 이름의 디스크 이미지가 표준 Apple
인스톨러 패키지를 포함하고 있습니다. 이것을 다운로드 받아 실행시키면, 설치프로그램은 비개발자 배포판non-developer
release)을 설치하게 될 것입니다.RAqua 는 네이브(naive) Mac OS X Darwin 버전의 R과 R.app Mac OS
X <acronym>GUI</acronym>를 지칭합니다.  <samp>bin/macosx/powerpc/contrib/<var>x</var>.<var>y</var></samp>
디렉토리 내에는, R의 &ldquo;<var>x</var>.<var>y</var>&rdquo; 릴리즈에 대응되는 RAqua 와 함께 사용될 (Mac OS X 의
powerpc 버전을 위하여) 미리 빌드된 바이너리 패키지들이 있을 것입니다. 이러한 패키지들의 설치는 R.app
<acronym>GUI</acronym>에 있는 &ldquo;Package&rdquo; 메뉴를 통하여 가능합니다.  Mac OS X 에 대한 이러한 R 포트 (port)
는 <a href="mailto:Stefano.Iacus@R-project.org">Stefano Iacus</a>에 의하여 유지 보수 됩니다. 더 자세한
사항에 대해서는 <a href="http://CRAN.R-project.org/bin/macosx/RMacOSX-FAQ.html">&ldquo;R for
Mac OS X <acronym>FAQ</acronym></a> 를 참고하길 바랍니다.
</p>
<p>Snapshots of the r-patched and r-devel versions of R are available as Apple
installer packages at <a href="http://r.research.att.com">http://r.research.att.com</a>.
</p>
<hr>
<a name="Are-there-Unix_002dlike-binaries-for-R_003f"></a>
<a name="Are-there-Unix_002dlike-binaries-for-R_003f-1"></a>
<h3 class="section">2.6 Are there Unix-like binaries for R?</h3>


<p>The <samp>bin/linux</samp> directory of a <acronym>CRAN</acronym> site contains the following
packages.
</p>
<blockquote>
<table>
<thead><tr><th></th><th>CPU</th><th>Versions</th><th>Provider</th></tr></thead>
<tr><td>Debian</td><td>i386/amd64</td><td>squeeze/wheezy</td><td>Johannes Ranke</td></tr>
<tr><td></td><td>armel</td><td>wheezy</td><td>Johannes Ranke</td></tr>
<tr><td>Ubuntu</td><td>i386/amd64</td><td>lucid/precise/trusty</td><td>Michael Rutter</td></tr>
</table>
</blockquote>

<p>Dirk Eddelbuettel에 의해 유지보수되는 데비안(Debian) 패지들은 오랜 시간에 걸쳐 데비안 배포판의 일부분이 되어가고
있으며, 데비안 패키지 관리도구인 APT 를 통하여 쉽게 엑세스 될 수 있습니다. R 기본 환경 및 추천 패키지들을 설치하기 위해서는
<code>apt-get install r-base r-recommended</code>을 입력하면 됩니다.  만약, 원시코드(source)로부터
R 패키지를 빌드하고자 한다면, 이에 필요한 도구들을 설치하기 위해서 <code>apt-get install r-base-dev</code>이라고
입력하면 됩니다.  최소한 <em>stable</em>(안정적인 버전의) 데비안에 적용이 될 수 있는 &ldquo;backport&rdquo;라고 불리우는 R
패키지들은 Johannes Ranke에 의하여 제공되며, <acronym>CRAN</acronym>으로부터 받을 수 있습니다.  R 데비안 패키지들에 대한 사항들과
백포트(backport)를 설치하는 하는 방법에 대해서는
<a href="http://CRAN.R-project.org/bin/linux/debian/README">http://CRAN.R-project.org/bin/linux/debian/README</a>를 살펴보시길 바랍니다.  이는
데비안으로부터 파생된 다른 배포판에서도 안정적으로 작동될 것입니다.  우분투를 위한 네이티브 백포트 (native backports)는
Michael Rutter에 의해서 제공됩니다.
</p>
<p>Fedora(페도라)를 위한 R 바이너리들은 Tom &ldquo;Spot&rdquo; Callaway에 의해서 유지보수되며, 페도라 배포판의 일부로서
제공되며 RPM 설치 및 업데이트 관리프로그램인 <code>yum</code>을 통하여 엑세스 할 수 있습니다.  페도라 R RPM 은
사용자와 개발자에게 필요한 모든 구성요소들을 다 함께 설치해주는 메타패키지(&ldquo;meta-package&rdquo;)인데, 이는
<code>R-core</code>와 <code>R-devel</code>으로 구분되어집니다.  또한, 독립적인 (standalone) R math
library (<code>libRmath</code>와 <code>libRmath-devel</code>)를 설치할 수 있으며, R 패키지들을 선택하여
설치하는 기능 역시 포함하고 있습니다.  Enterprise Linux (EPEL) 프로젝트
(<a href="http://fedoraproject.org/wiki/EPEL">http://fedoraproject.org/wiki/EPEL</a>)를 위한 추가적인 패키지들은 RedHat Enterprise
Linux와 이와 호환되는 배포판들을 위한 Fedora RPM의 포트를 제공합니다.  새로운 버전의 R이 배포될때, 페도라 RPM 에서
새버전을 사용하기 위해서는 최대 2주까지 지연될 가능성이 있습니다. 이는 반드시 페도라내에 지정된 검증프로세스 (statuory
Fedora review process)를 반드시 거쳐야 하기 때문입니다.
</p>
<p>See <a href="http://CRAN.R-project.org/bin/linux/suse/README.html">http://CRAN.R-project.org/bin/linux/suse/README.html</a> for
information about RPMs for openSUSE.
</p>
<p>No other binary distributions are currently publically available via
<acronym>CRAN</acronym>.
</p>
<hr>
<a name="What-documentation-exists-for-R_003f"></a>
<a name="What-documentation-exists-for-R_003f-1"></a>
<h3 class="section">2.7 What documentation exists for R?</h3>

<p>R에서 사용할 수 있는 대부분의 함수들과 변수들에 대해서는 온라인 문서가 있으며, 이들은 R 프롬프트에서
<kbd>help(<var>name</var>)</kbd> (또는 <kbd>?<var>name</var></kbd>)이라고 입력함으로서 스크린을 통하여 확인할 수
있습니다.  여기에서 <var>name</var>은 도움을 받고자 하는 주제를 의미합니다.  (단항과 이항 연산자 (unary and binary
operators), 그리고 분기문(control-flow)과 같은 특수한 형태의 경우에는 검색어를 입력할때 따옴표를 함께 넣어주어야 할
수도 있습니다).
</p>
<p>본 문서 또한 <acronym>HTML</acronym>과 PDF 형식으로 온라인으로 읽을 수 있는 하나의 참고메뉴얼(reference manual)이며,
LaTeX를 통하여 출력할 수도 있습니다.  <a href="#How-can-R-be-installed_003f">How can R be installed?</a>를 참고하시길 바랍니다.  가장
최신버전의 <acronym>HTML</acronym> 버전은 항상 <a href="http://stat.ethz.ch/R-manual/">http://stat.ethz.ch/R-manual/</a>에서 확인해 보실수 있습니다.
</p>
<p>일부 버전들에 대한 참고매뉴얼 (R reference manual)들은 Network Theory Ltd
<a href="http://www.network-theory.co.uk/R/base/">http://www.network-theory.co.uk/R/base/</a>에서 구매할 수 있습니다.  매뉴얼 한 부가 판매될
때마다, 출판사는 R Foundation (see <a href="#What-is-the-R-Foundation_003f">What is the R Foundation?</a>)에 USD 10 (미국돈으로
10달러)를 기부합니다.
</p>
<p>R은 다음과 같은 매뉴얼들과 함께 배포되고 있습니다.
</p>
<ul>
<li> &ldquo;An Introduction to R&rdquo; (<samp>R-intro</samp>)
는 데이터의 유형, 프로그래밍 구성요소들, 통계 모델링과 그래픽스에 대한 정보를 포함하고 있습니다. 이 문서는 Bill Venables과
David Smith가 작성한 &ldquo;Notes on S-Plus&rdquo;라는 문서를 토대로 하고 있습니다.
</li><li> &ldquo;Writing R Extensions&rdquo; (<samp>R-exts</samp>)
는 R 애드온(add-on) 패키지의 작성절차, R 문서 작성요령, R과 다른 언어들과의 인터페이스, 그리고 <acronym>API</acronym>에
대해서 기술하고 있습니다.
</li><li> &ldquo;R Data Import/Export&rdquo; (<samp>R-data</samp>)
는 데이터 입출력에 대한 가이드입니다.
</li><li> &ldquo;The R Language Definition&rdquo; (<samp>R-lang</samp>),
는 &ldquo;Kernighan &amp; Ritchie of R&rdquo;이라는 문서의 최초 버전이기도 하며, 이밸류에이션(evaluation, 평가,
파싱(parsing, 구문해석), 객체지향 프로그래밍 (object oriented programming), 연산(computing)
등의 내용을 다루고 있습니다.
</li><li> &ldquo;R Installation and Administration&rdquo; (<samp>R-admin</samp>).
</li><li> &ldquo;R Internals&rdquo; (<samp>R-ints</samp>)
R의 내부구조(internal structure)에 대한 가이드입니다 (R-2.4.0에서 추가되었습니다).
</li></ul>

<p>R과 관련된 문헌(publications)들에 대한 정보는 BibTeX 형식으로 아래의 주소에서 찾아 볼 수있습니다.
</p>
<div class="display">
<pre class="display"><a href="http://www.R-project.org/doc/bib/R.bib">http://www.R-project.org/doc/bib/R.bib</a>
</pre></div>

<p>R 코어팀 멤버들에 의해서 씌여진 책들은 아래와 같습니다.
</p>
<blockquote>
<p>John M. Chambers (2008), &ldquo;Software for Data Analysis: Programming with
R&rdquo;.  Springer, New York, ISBN 978-0-387-75935-7,
<a href="http://stat.stanford.edu/~jmc4/Rbook/">http://stat.stanford.edu/~jmc4/Rbook/</a>.
</p>
<p>Peter Dalgaard (2008), &ldquo;Introductory Statistics with R&rdquo;, 2nd edition.
Springer, ISBN 978-0-387-79053-4,
<a href="http://www.biostat.ku.dk/~pd/ISwR.html">http://www.biostat.ku.dk/~pd/ISwR.html</a>.
</p>
<p>Robert Gentleman (2008), &ldquo;R Programming for Bioinformatics&rdquo;.  Chapman &amp;
Hall/CRC, Boca Raton, FL, ISBN 978-1-420-06367-7,
<a href="http://www.bioconductor.org/pub/RBioinf/">http://www.bioconductor.org/pub/RBioinf/</a>.
</p>
<p>Stefano M. Iacus (2008), &ldquo;Simulation and Inference for Stochastic
Differential Equations: With R Examples&rdquo;. Springer, New York, ISBN
978-0-387-75838-1.
</p>
<p>Deepayan Sarkar (2007), &ldquo;Lattice: Multivariate Data Visualization with
R&rdquo;. Springer, New York, ISBN 978-0-387-75968-5.
</p>
<p>W. John Braun and Duncan J. Murdoch (2007), &ldquo;A First Course in Statistical
Programming with R&rdquo;.  Cambridge University Press, Cambridge, ISBN
978-0521872652.
</p>
<p>P. Murrell (2005), &ldquo;R Graphics&rdquo;, Chapman &amp; Hall/CRC, ISBN: 1-584-88486-X,
<a href="http://www.stat.auckland.ac.nz/~paul/RGraphics/rgraphics.html">http://www.stat.auckland.ac.nz/~paul/RGraphics/rgraphics.html</a>.
</p>
<p>William N. Venables and Brian D. Ripley (2002), &ldquo;Modern Applied Statistics
with S&rdquo; (4th edition).  Springer, ISBN 0-387-95457-0,
<a href="http://www.stats.ox.ac.uk/pub/MASS4/">http://www.stats.ox.ac.uk/pub/MASS4/</a>.
</p>
<p>Jose C. Pinheiro and Douglas M. Bates (2000), &ldquo;Mixed-Effects Models in S
and S-Plus&rdquo;. Springer, ISBN 0-387-98957-0.
</p></blockquote>

<p>마지막으로, 최근의 것은 아니지만, Ross와 Robert가 R을 디자인하고 구현하면서 겪은 경험이 다음의 문서에 기록되어 있습니다.
Ihaka &amp; Gentleman (1996), &ldquo;R: A Language for Data Analysis and Graphics&rdquo;,
<a href="http://www.amstat.org/publications/jcgs/"><em>Journal of
Computational and Graphical Statistics</em></a>, <strong>5</strong>, 299&ndash;314.
</p>
<hr>
<a name="Citing-R"></a>
<a name="Citing-R-1"></a>
<h3 class="section">2.8 Citing R</h3>

<p>R을 참고자료로서 인용하고자 한다면, 아래의 BibTeX를 이용하시면 됩니다.
</p>
<div class="example">
<pre class="example">@Manual{,
  title        = {R: A Language and Environment for Statistical
                  Computing},
  author       = {{R Core Team}},
  organization = {R Foundation for Statistical Computing},
  address      = {Vienna, Austria},
  year         = 2014,
  url          = {http://www.R-project.org}
}
</pre></div>

<p>R과 R 패키지들에 대한 BibTeX 항목들또는 인용정보들은 <code>citation()</code>을 통하여 얻을 수 있습니다.
</p>
<hr>
<a name="What-mailing-lists-exist-for-R_003f"></a>
<a name="What-mailing-lists-exist-for-R_003f-1"></a>
<h3 class="section">2.9 What mailing lists exist for R?</h3>

<p>이러한 서비스를 제공해 준 <a href="mailto:Martin.Maechler@R-project.org">Martin Maechler</a>에게 감사를
드립니다. 네가지 종류의 메일링 리스트가 있습니다.
</p>
<dl compact="compact">
<dt><code>R-announce</code></dt>
<dd><p>는 R의 개발과 새로운 코드의 유용성에 관한 주요 공지사항들을 위한 (moderated) 리스트입니다.
</p></dd>
<dt><code>R-packages</code></dt>
<dd><p>는 새롭게 개발되거나 주요기능들이 강화된 패키지들의 유용성에 대한 공지사항들을 위한 (moderated) 리스트입니다.
</p></dd>
<dt><code>R-help</code></dt>
<dd><p>은 R을 사용하면서 겪는 문제들과 이들에 대한 해법들을 논의하는 목적으로 사용되는 주된 메일링 리스트입니다.  R의 개발과 새로운
코드들의 유용성에 관하여 &lsquo;R-announce&rsquo;와 &lsquo;R-packages&rsquo;에서 다루지 않은 공지사항들을 포함하기도 하며, R의 원시코드와
문서에 대한 강화된 기능 및 패치들에 대한 내용을 다루기도 합니다.  S와 <small>S-PLUS</small>와의 호환성과 비교에 대한 내용도 있으며, 좋은
예제들과 벤치마크들을 포스팅할 목적으로도 사용합니다.
</p></dd>
<dt><code>R-devel</code></dt>
<dd><p>는 R의 코드개발 (code development)에 관한 질문들과 논의를 위한 리스트입니다.  R의 미래에 대한 논의, 새로운 기능에
대한 제안, 그리고 새로운 버전에 대한 테스트를 목적으로 이용될 수 있습니다.  이 리스트는 R의 개발에 적극적인 자세 (active
position)을 유지하는 사람을 위한 것입니다.
</p></dd>
</dl>

<p>위에서 나열한 어떠한 종류의 메일링 리스트에 어떤 내용을 보내기 <em>전에 반드시</em>
<a href="http://www.R-project.org/posting-guide.html">posting guide(포스팅 가이드)</a>를
읽어보시길 바랍니다.
</p>
<p>R-help 는 본래 R을 이용하여 본인의 문제를 해결하고자 하지만 프로그래밍에 대한 지식이 부족한 분들을 위해서 개설되었다는 것을
아셨으면 합니다. 따라서, 프로그래머가 아닌 사람들이 이해하기 어려운 질문들 (즉, C 혹은 C++ 과 연관된 내용)은 R-devel을
이용하시길 부탁드립니다.  
</p>
<p>이러한 메일링 리스트들에 대한 구독 및 보관된 자료들에 대한 열람은
<a href="http://stat.ethz.ch/mailman/listinfo/">http://stat.ethz.ch/mailman/listinfo/</a>을 통하여 손쉽게 하실 수 있습니다. 이는
<a href="mailto:R-help-request@lists.R-project.org">R-help-request@lists.R-project.org</a>에 이메일을 보낼 때 subject(제목)이 아닌
<em>body(본문)</em>에 &lsquo;<samp>subscribe(구독)</samp>&rsquo; (또는 &lsquo;<samp>unsubscribe(구독해지)</samp>&rsquo;라고 이메일을
보냄으로서 이용하실 수 있습니다.
</p>
<p>R-help 메일링 리스트를 구독하는 모든 이들에게 메시지를 보내기 위해서는
<a href="mailto:R-help@lists.R-project.org">R-help@lists.R-project.org</a>로 이메일을 보내주세요. 다른 메일링 리스트들의 구독 및 포스팅 요령은
r-help와 동일합니다.  &lsquo;<samp>R-help</samp>&rsquo; 대신에 &lsquo;<samp>R-announce</samp>&rsquo;, &lsquo;<samp>R-packages</samp>&rsquo;,
&lsquo;<samp>R-devel</samp>&rsquo; 이라고만 변경해주시면 됩니다. R-announce과 R-packages는 R-help에 이르는 과정이라는 점을
아시길 바랍니다.  만약 R-help를 구독하고 있지 않다면 둘 중 어느 하나는 구독을 해야합니다.
</p>
<p><strong>사용자가 R 코어 개발자들에게 개별적으로 이메일을 보내기 보다는 R-help에 메일을 보내주시길 부탁드립니다 (코어 멤버들
역시 메일링을 사용하고 있습니다)</strong>.  이러한 부탁을 하는 이유는 코어 개발자들이 지속적으로 R을 향상시키는데 있어서 많은 시간을
줄여주기도 할 뿐더러, 질문자가 가장 빠른 시간내에 답변을 받을 수 있기 때문입니다. 
</p>
<p>버그리포트를 할 때에는 버그를 정확하게 재구현할 수 있도록 코드를 함께 올려주시면 매우 도움이 됩니다.  또한, 이러한 버그를 보고할 때
사용자가 어떤 시스템과 어떤 버전의 R 을 사용하는지 알려주시길 부탁드립니다. 더 자세한 사항들에 대해서는 <a href="#R-Bugs">R Bugs</a>를
살펴보세요.
</p>
<p>R 메일링 리스트들에 대한 좀 더 많은 정보를 알고 싶으시다면 <a href="http://www.R-project.org/mail.html">http://www.R-project.org/mail.html</a>을
살펴보시길 바랍니다.
</p>
<p>코멘트와 리포트들을 <a href="mailto:R-core@lists.R-project.org">R-core@lists.R-project.org</a>을 통하여 R 코어팀에 보내주실 수 있습니다.
</p>
<p>R project의 메일링 리스트들의 대부분은 또한 NNTP 뉴스리더를 이용하는 웹브라우저를 이용하여 읽을 수 있는
<a href="http://gmane.org">Gmane</a> 또는 RSS 피드를 통해서 이용할 수 있습니다.  사용가능한 메일링 리스트들에
대해서는 <a href="http://dir.gmane.org/index.php?prefix=gmane.comp.lang.r.">http://dir.gmane.org/index.php?prefix=gmane.comp.lang.r.</a>에서
확인하시길 바라며, RSS 피드에 대한 더 자세한 사항들은 <a href="http://www.gmane.org/rss.php">http://www.gmane.org/rss.php</a>에서
확인부탁드립니다.
</p>
<hr>
<a name="What-is-CRAN_003f"></a>
<a name="What-is-CRAN_003f-1"></a>
<h3 class="section">2.10 What is <acronym>CRAN</acronym>?</h3>

<p>&ldquo;Comprehensive R Archive Network&rdquo; (<acronym>CRAN</acronym>)은 R 배포판, 기여된 확장 프로그램들 (즉,
패키지들), R과 관련된 문서들, 그리고 바이너리들로 구성되어 있는 동일한 자료들을 전달하는 사이트들의 모음을 의미합니다.
</p>
<p>다음의 <acronym>URL</acronym>로부터 오스트리아(Austria) 빈(Wien) Wirtschaftsuniversit에서 <acronym>CRAN</acronym>
마스터 사이트를 찾을 수 있습니다.
</p>
<blockquote>
<p><a href="http://CRAN.R-project.org/">http://CRAN.R-project.org/</a>
</p></blockquote>

<p>데일리 미러(daily mirrors)들은 <acronym>URL</acronym>에서 찾을 수 있습니다.
</p>
<blockquote>
<table>
<tr><td width="45%"><a href="http://cran.at.R-project.org/">http://cran.at.R-project.org/</a></td><td width="40%">(WU Wien, Austria)</td></tr>
<tr><td width="45%"><a href="http://cran.au.R-project.org/">http://cran.au.R-project.org/</a></td><td width="40%">(PlanetMirror, Australia)</td></tr>
<tr><td width="45%"><a href="http://cran.br.R-project.org/">http://cran.br.R-project.org/</a></td><td width="40%">(Universidade Federal do Paran&aacute;, Brazil)</td></tr>
<tr><td width="45%"><a href="http://cran.ch.R-project.org/">http://cran.ch.R-project.org/</a></td><td width="40%">(ETH Z&uuml;rich, Switzerland)</td></tr>
<tr><td width="45%"><a href="http://cran.dk.R-project.org/">http://cran.dk.R-project.org/</a></td><td width="40%">(SunSITE, Denmark)</td></tr>
<tr><td width="45%"><a href="http://cran.es.R-project.org/">http://cran.es.R-project.org/</a></td><td width="40%">(Spanish National Research Network, Madrid, Spain)</td></tr>
<tr><td width="45%"><a href="http://cran.fr.R-project.org/">http://cran.fr.R-project.org/</a></td><td width="40%">(INRA, Toulouse, France)</td></tr>
<tr><td width="45%"><a href="http://cran.pt.R-project.org/">http://cran.pt.R-project.org/</a></td><td width="40%">(Universidade do Porto, Portugal)</td></tr>
<tr><td width="45%"><a href="http://cran.uk.R-project.org/">http://cran.uk.R-project.org/</a></td><td width="40%">(U of Bristol, United Kingdom)</td></tr>
<tr><td width="45%"><a href="http://cran.za.R-project.org/">http://cran.za.R-project.org/</a></td><td width="40%">(Rhodes U, South Africa)</td></tr>
</table>
</blockquote>

<p>미러들의 모든 목록에 대해서는 <a href="http://CRAN.R-project.org/mirrors.html">http://CRAN.R-project.org/mirrors.html</a> 를 참고해주세요.
네트워크 부담을 줄이기 위해서 현재 접속하고 계신 위치에서 가장 가까운 위치에 있는 <acronym>CRAN</acronym> 사이트를 이용해주세요.
</p>
<p><acronym>CRAN</acronym>으로부터 R의 가장 최신 공식 릴리즈, R의 데일리 스냅샷 (현재 소스트리의 복사본), gzip과 bzip으로 압축된 tar
파일들, 추가적으로 기여된 코드들, Linux, Mac OS Classic, Mac OS X, 그리고 MS Windows와 같은 다양한
운영체제를 위하여 미리 빌드된 바이너리들을 얻을 수 있습니다.  <acronym>CRAN</acronym>은 또한 R 문서들, 메일링 리스트들, 그리고 버그트랙킹
시스템에 접근할 수 있도록 해줍니다.
</p>
<p><acronym>CRAN</acronym>에 &ldquo;submit&rdquo;(제출)하기 위해서는 사용자의 제출(submission)이
<a href="http://CRAN.R-project.org/web/packages/policies.html">CRAN Repository
Policy</a>에 부합하는지 먼저 확인을 해야 합니다. 
</p>
<hr>
<a name="Can-I-use-R-for-commercial-purposes_003f"></a>
<a name="Can-I-use-R-for-commercial-purposes_003f-1"></a>
<h3 class="section">2.11 Can I use R for commercial purposes?</h3>

<p>R은 <a href="http://www.gnu.org/licenses/old-licenses/gpl-2.0.html"><acronym>GNU</acronym> General Public License (<acronym>GPL</acronym>), version 2</a>하에 배포됩니다.만약,
어떤 특정한 상황에서 R을 사용하는 것이 합법적인가에 대한 의구심이 생길 경우, 반드시 전문 법률상담가와 상의하시길 바랍니다. 우리는
법적 조언을 줄 수 있는 그 어떠한 위치에 있지 않습니다.
</p>
<p>R Core Team은 R 을 상업적 목적 (즉, 비지니스 혹은 컨설팅)으로도 사용할 수 있다고 생각합니다.  모든 오픈소스의
라이센스들과 같이 <acronym>GPL</acronym>은 패키지들의 모든 그리고 어떠한 (all and any) 용도로 사용되는 것을 허용합니다.
이것은 오로지 R 혹은 R 코드의 일부분을 포함하는 프로그램의 배포를 제한하는
것입니다. <a href="http://www.opensource.org/docs/definition.html">Open Source
Definition</a>의 제 6항 (clause 6) &ldquo;No Discrimination Against Fields of
Endeavor&rdquo;에서 명확하게 나타납니다.
</p>
<blockquote>
<p>The license must not restrict anyone from making use of the program in a
specific field of endeavor.  For example, it may not restrict the program
from being used in a business, or from being used for genetic research.
</p></blockquote>

<p>이는 또한 GPL 제 0항 (clause 0)에서도 아래와 같이 (일부분만) 언급되어 있습니다. 
</p>
<blockquote>
<p>모든 추천(recommended) 패키지들을 포함하여 대부분의 애드온(add-on) 패키지들은 이러한 방법으로서 상업적으로도 이용이 될
수 있음을 명시하고 있습니다.  몇개의 패키지들은 &ldquo;non-commercial use&rdquo;(비상업용적 용도)로서 제한됩니다.  이 경우
사용자는 반드시 저작권자에게 상업적 용도로의 활용여부를 문의하거나 법률상담가로부터의 조언을 찾아야 합니다.
</p></blockquote>

<p>Most add-on packages, including all recommended ones, also explicitly allow
commercial use in this way.  A few packages are restricted to
&ldquo;non-commercial use&rdquo;; you should contact the author to clarify whether
these may be used or seek the advice of your legal counsel.
</p>
<p>이 섹션에 언급된 모든 논의는 법적조언이 아닙니다. R Core Team은 어떠한 경우에도 법적조언을 제공하지 않습니다.
</p>
<hr>
<a name="Why-is-R-named-R_003f"></a>
<a name="Why-is-R-named-R_003f-1"></a>
<h3 class="section">2.12 Why is R named R?</h3>

<p>R이라는 이름은 R을 최초로 개발한 두 개발자인 Robert Gentleman과 Ross Ihaka의 이름에서 기인하기도 하며, 동시에
Bell 연구소의 &rsquo;S&rsquo; 라는 언어에서 영감을 받기도 하였습니다 (see <a href="#What-is-S_003f">What is S?</a> 섹션을 참고하세요).
</p>

<hr>
<a name="What-is-the-R-Foundation_003f"></a>
<a name="What-is-the-R-Foundation_003f-1"></a>
<h3 class="section">2.13 What is the R Foundation?</h3>

<p>The R Foundation은 공익(public interest)을 위한 비영리단체(not for profit
organization)입니다. 이 단체는 R project와 통계계산 (statistical computing)에 연관된 기술들을
지원하고, R 개발 커뮤니티(development community)를 지원하거나 함께 일하고 싶어하는 개인, 기관 및 영리추구의
사업체(commercial enterprises)들에게 기준(reference point)을 제공하며, R 소프트웨어와 문서에 대한
저작권을 관리 및 유지하기 위하여 R Core Team 멤버들에 의하여
설립되었습니다. <a href="http://www.R-project.org/foundation/">http://www.R-project.org/foundation/</a>에서 더 많은 정보를 찾을 수 있습니다.
</p>
<hr>
<a name="What-is-R_002dForge_003f"></a>
<a name="What-is-R_002dForge_003f-1"></a>
<h3 class="section">2.14 What is R-Forge?</h3>

<p>R-Forge (<a href="http://R-Forge.R-project.org/">http://R-Forge.R-project.org/</a>)는 R 패키지, R 관련 소프트웨어, 그리고 다양한
프로젝트들의 개발을 위한 중요한 기반(central platform)을 제공합니다.  SVN에 최적화된 손쉬운 접근, 매일 빌드되고
체크되는 패키지들, 메일링 리스트, 버그 트래킹, 메시지 보드/포럼공간, 사이트 호스팅, 영구파일 저장공간, 안정적인 백업, 그리고
웹기반의 관리 기능등을 제공하는 <a href="http://www.gforge.org/">GForge</a>를 기초로 운영됩니다.  R-Forge
웹페이지와 Stefan Theu&szlig;l and Achim Zeileis (2009), &ldquo;Collaborative software
development using R-Forge&rdquo;, <a href="http://journal.R-project.org/"><em>The
R Journal</em></a>, <strong>1</strong>(1):9&ndash;14 로부터 더 많은 내용을 얻을 수 있습니다.
</p>

<hr>
<a name="R-and-S"></a>
<a name="R-and-S-1"></a>
<h2 class="chapter">3 R and S</h2>


<hr>
<a name="What-is-S_003f"></a>
<a name="What-is-S_003f-1"></a>
<h3 class="section">3.1 What is S?</h3>

<p>S는 데이터 분석과 그래픽스를 위한 하이레벨 언어 (high level language)이며 환경입니다. 1998년에
Association for Computing Machinery (<acronym>ACM</acronym>)는 1998년에 S언어의 주요설계자인 John
M. Chambers에게 아래와 같이 치하하며 Software System Award를 수여하였습니다.
</p>
<blockquote>
<p>the S system, which has forever altered the way people analyze, visualize,
and manipulate data &hellip;
</p>
<p>S 언어의 발전은 John Chambers와 공동저자들에 의하여 쓰여진 네권의 책들에 잘 설명되어 있습니다.  이들은 또한 S에 대한
주요 참고문헌이기도 합니다.
</p></blockquote>

<p>The evolution of the S language is characterized by four books by John
Chambers and coauthors, which are also the primary references for S.
</p>
<ul>
<li> Richard A. Becker and John M. Chambers (1984), &ldquo;S.  An Interactive
Environment for Data Analysis and Graphics,&rdquo; Monterey: Wadsworth and
Brooks/Cole.

<p>이 책은 &ldquo;<em>Brown Book</em>&rdquo;(브라운 북)이라고도 알려져 있으며, 역사적 관심으로서입니다.
</p>
</li><li> Richard A. Becker, John M. Chambers and Allan R. Wilks (1988), &ldquo;The New S
Language,&rdquo; London: Chapman &amp; Hall.

<p>이 책은 &ldquo;<em>Blue Book</em>&rdquo;(블루 북)이라고도 하며, S 버전 2로 알려진 것이 무엇인지를 소개합니다.
</p>
</li><li> John M. Chambers and Trevor J. Hastie (1992), &ldquo;Statistical Models in S,&rdquo;
London: Chapman &amp; Hall.

<p>이 책은 <em>White Book</em>라고도 하며, S에서 통계모델링(statistical modeling)의 기능을 수행하기 위한
구조들이 추가되어진 S 버전 3로 알려진 것을 소개합니다. 
</p>
</li><li> John M. Chambers (1998), &ldquo;Programming with Data,&rdquo; New York: Springer, ISBN
0-387-98503-4 (<a href="http://cm.bell-labs.com/cm/ms/departments/sia/Sbook/">http://cm.bell-labs.com/cm/ms/departments/sia/Sbook/</a>).

<p>&ldquo;<em>Green Book</em>&rdquo;(그린 북)은 프로그래밍 프로세스의 매 단계마다 유용성을 향상시킬 수 있도록 John
Chambers에 의하여 설계된 S의 주요수정사항들이 있는 S 버전 4를 설명합니다.
</p></li></ul>

<p>더 자세한 내용은
<a href="http://cm.bell-labs.com/cm/ms/departments/sia/S/history.html">http://cm.bell-labs.com/cm/ms/departments/sia/S/history.html</a>에서
&ldquo;Stages in the Evolution of S&rdquo;를 살펴보시길 바랍니다.
</p>
<p><acronym>CMU</acronym>에 있는 <a href="http://lib.stat.cmu.edu/S/">S Repository</a>에 가시면 사용자들이
제공한 매우 방대한 양의 S코드들을 있습니다.
</p>

<hr>
<a name="What-is-S_002dPLUS_003f"></a>
<a name="What-is-S_002dPLUS_003f-1"></a>
<h3 class="section">3.2 What is <small>S-PLUS</small>?</h3>

<p><small>S-PLUS</small>는 <a href="http://www.tibco.com/">TIBCO Software Inc</a>에서 2008년에 인수한
<a href="http://www.insightful.com">Insightful Corporation</a>에서 판매하는 S의
부가가치된(value-added) 버전입니다 (즉, 상업용입니다).  더 많은 정보를 원하시면
<a href="http://www.insightful.com/products/splus/">Insightful <small>S-PLUS</small>
page</a>페이지와 <a href="http://spotfire.tibco.com/Products/S-Plus-Overview.aspx">TIBCO Spotfire S+ Products page</a>제품 페이지를 참조하세요.
</p>
<hr>
<a name="What-are-the-differences-between-R-and-S_003f"></a>
<a name="What-are-the-differences-between-R-and-S_003f-1"></a>
<h3 class="section">3.3 What are the differences between R and S?</h3>

<p>S는 현재 세가지 종류의 구현체 또는 &ldquo;engines&rdquo;(엔진)을 가진 언어로서 여길 수 있습니다.  이는 &ldquo;old S
engine&rdquo; (S version 3; S-Plus 3.x 그리고 4.x), &ldquo;new S engine&rdquo; (S version 4;
S-Plus 5.x 혹은 그 이후의 버전), 그리고 R입니다.  이러한 내용을 바탕으로 보면, &ldquo;R과 S 사이의 다른점&rdquo;에 대한
질문은 S언어를 이용하여 구현된 R의 세부사항들을 묻는것과 같습니다. 즉, R의 엔진과 S의 엔진 사이에 다른 점의 다른 점입니다.
</p>
<p>For the remainder of this section, &ldquo;S&rdquo; refers to the S engines and not the
S language.
</p>

<hr>
<a name="Lexical-scoping"></a>
<a name="Lexical-scoping-1"></a>
<h4 class="subsection">3.3.1 Lexical scoping</h4>

<p>S언어를 이용한 다른 구현체들과는 다르게 R은 내포된 함수 정의들(nested function definitions)이 렉시컬
스코프(lexically scoped)된 평가모델(evaluation model)을 사용합니다.  이것은 Scheme의 평가모델과
유사합니다.
</p>
<p>이 차이점은 함수(function)내에서 <em>free</em>(자유)변수들이 발생할때 분명해집니다. 자유변수들은 함수의 인자목록에서
사용된 형식인자(formal argument)도 아니고, 함수의 본체에서 이들에 할당함으로서 생성된 지역변수(local
variables)도 아닙니다. S에서는 자유변수들은 C와 유사하게 전역변수(global variable)들의 묶음에 의해서 결정되어
집니다.  (C에서는 지역적 범위와 전역적 범위만이 있습니다).  R에서 이들은 함수가 생성되어진 환경(environment)에 의해서
결정되어집니다.
</p>
<p>다음의 함수를 보십시오:
</p>
<div class="example">
<pre class="example">cube &lt;- function(n) {
  sq &lt;- function() n * n
  n * sq()
}
</pre></div>

<p>S에서, <code>sq()</code>는 변수 <code>n</code>이 전역적으로 정의되지 않는 이상은 이에 대해서 알지 못합니다:
</p>
<div class="example">
<pre class="example">S&gt; cube(2)
Error in sq():  Object &quot;n&quot; not found
Dumped
S&gt; n &lt;- 3
S&gt; cube(2)
[1] 18
</pre></div>

<p>R에서는 <code>cube()</code>가 실행되었을때 생성된 &ldquo;environment&rdquo;(환경)이 살펴지게 됩니다:
</p>
<div class="example">
<pre class="example">R&gt; cube(2)
[1] 8
</pre></div>







<p>실제로 다루어 볼 수 있는 흥미로운 문제로서, (연속)분포에서 샘플의 크기 <em>n</em>으로부터 <em>r</em> 번째의 order
statistic (순서통계량)의 밀도함수를 구하는 함수를 작성해본다고 가정해 봅니다.  단순한 방법으로 우리는 분포의 cdf와 pdf
모두를 explicit arguments(명시적 인자들)로 사용해야 합니다.  (이는 Luke Tierney에 의하여 작성된 다양한
포스팅으로부터 뽑아낸 예제입니다).
</p>
<p><code>call()</code>에 대한 <small>S-PLUS</small> 문서는 기본적으로 다음과 같이 하도록 안내합니다.
</p>
<div class="example">
<pre class="example">dorder &lt;- function(n, r, pfun, dfun) {
  f &lt;- function(x) NULL
  con &lt;- round(exp(lgamma(n + 1) - lgamma(r) - lgamma(n - r + 1)))
  PF &lt;- call(substitute(pfun), as.name(&quot;x&quot;))
  DF &lt;- call(substitute(dfun), as.name(&quot;x&quot;))
  f[[length(f)]] &lt;-
    call(&quot;*&quot;, con,
         call(&quot;*&quot;, call(&quot;^&quot;, PF, r - 1),
              call(&quot;*&quot;, call(&quot;^&quot;, call(&quot;-&quot;, 1, PF), n - r),
                   DF)))
  f
}
</pre></div>

<p>Rather tricky, isn&rsquo;t it?  The code uses the fact that in S,
이 코드는 S에서는 함수들이 마지막 인자와 같이 함수의 본체의 특정한 방식(mode)의 목록들이라는 사실을 이용하고 있습니다.  이러한
방식은 R에서 작동하지 않습니다 (누군가는 여기에서 사용된 아이디어를 성공적으로 구현할 수도 있을것입니다).
</p>
<p><code>substitute()</code>의 사용빈도가 매우 높으며 S와 R 모두 작동하는 것처럼 보이는 버전은 다음과 같습니다. 
</p>
<div class="example">
<pre class="example">dorder &lt;- function(n, r, pfun, dfun) {
  con &lt;- round(exp(lgamma(n + 1) - lgamma(r) - lgamma(n - r + 1)))
  eval(substitute(function(x) K * PF(x)^a * (1 - PF(x))^b * DF(x),
                  list(PF = substitute(pfun), DF = substitute(dfun),
                       a = r - 1, b = n - r, K = con)))
}
</pre></div>

<p>(S에서는 <code>eval()</code>이 필요하지 않습니다).
</p>
<p>그러나, R에서는 다음과 같은 더욱 쉬운 해결책이 있습니다:
</p>
<div class="example">
<pre class="example">dorder &lt;- function(n, r, pfun, dfun) {
  con &lt;- round(exp(lgamma(n + 1) - lgamma(r) - lgamma(n - r + 1)))
  function(x) {
    con * pfun(x)^(r - 1) * (1 - pfun(x))^(n - r) * dfun(x)
  }
}
</pre></div>

<p>이것은 &ldquo;natural(자연스러운)&rdquo; 구현 같이 보이며, 귀환된 함수(returned function)내에 있는 자유변수(free
variable)들이 환경을 규정한 내부에서 찾아질 수 있기 때문에 잘 실행됩니다 (이것을 렉시컬 스코프(lexical scope)라고
합니다). 
</p>
<p>꼭 알아야 할 것은 함수 <em>closure</em>라는 것입니다. 즉, 함수를 평가하기 위하여 필요한 모든 variable
binding(변수 바인딩)들 있는 본체를 의미합니다.  위의 버전의 경우, 값을 지닌 함수내의 자유변수들은 변경되지 않았기 때문에
S에서 이를 사용할 수 있을 뿐아니라, 만약 클로저 작동 (closure operation)을 함수 <code>MC()</code>로 이끌어 내고자
한다면 (클로져를 만들기 위해서) 다음과 같이 할 수 있습니다. 
</p>
<div class="example">
<pre class="example">dorder &lt;- function(n, r, pfun, dfun) {
  con &lt;- round(exp(lgamma(n + 1) - lgamma(r) - lgamma(n - r + 1)))
  MC(function(x) {
       con * pfun(x)^(r - 1) * (1 - pfun(x))^(n - r) * dfun(x)
     },
     list(con = con, pfun = pfun, dfun = dfun, r = r, n = n))
}
</pre></div>

<p>Given the appropriate definitions of the closure operator, this works in
both R and S, and is much &ldquo;cleaner&rdquo; than a substitute/eval solution (or
one which overrules the default scoping rules by using explicit access to
evaluation frames, as is of course possible in both R and S).
</p>
<p>For R, <code>MC()</code> simply is
</p>
<div class="example">
<pre class="example">MC &lt;- function(f, env) f
</pre></div>

<p>(lexical scope!), a version for S is
</p>
<div class="example">
<pre class="example">MC &lt;- function(f, env = NULL) {
  env &lt;- as.list(env)
  if (mode(f) != &quot;function&quot;)
    stop(paste(&quot;not a function:&quot;, f))
  if (length(env) &gt; 0 &amp;&amp; any(names(env) == &quot;&quot;))
    stop(paste(&quot;not all arguments are named:&quot;, env))
  fargs &lt;- if(length(f) &gt; 1) f[1:(length(f) - 1)] else NULL
  fargs &lt;- c(fargs, env)
  if (any(duplicated(names(fargs))))
    stop(paste(&quot;duplicated arguments:&quot;, paste(names(fargs)),
         collapse = &quot;, &quot;))
  fbody &lt;- f[length(f)]
  cf &lt;- c(fargs, fbody)
  mode(cf) &lt;- &quot;function&quot;
  return(cf)
}
</pre></div>

<p>Similarly, most optimization (or zero-finding) routines need some arguments
to be optimized over and have other parameters that depend on the data but
are fixed with respect to optimization.  With R scoping rules, this is a
trivial problem; simply make up the function with the required definitions
in the same environment and scoping takes care of it.  With S, one solution
is to add an extra parameter to the function and to the optimizer to pass in
these extras, which however can only work if the optimizer supports this.
</p>
<p>Nested lexically scoped functions allow using function closures and
maintaining local state.  A simple example (taken from Abelson and Sussman)
is obtained by typing <kbd>demo(&quot;scoping&quot;)</kbd> at the R prompt.  Further
information is provided in the standard R reference &ldquo;R: A Language for Data
Analysis and Graphics&rdquo; (see <a href="#What-documentation-exists-for-R_003f">What documentation exists for R?</a>) and in
Robert Gentleman and Ross Ihaka (2000), &ldquo;Lexical Scope and Statistical
Computing&rdquo;, <a href="http://www.amstat.org/publications/jcgs/"><em>Journal
of Computational and Graphical Statistics</em></a>, <strong>9</strong>, 491&ndash;508.
</p>
<p>Nested lexically scoped functions also imply a further major difference.
Whereas S stores all objects as separate files in a directory somewhere
(usually <samp>.Data</samp> under the current directory), R does not.  All objects
in R are stored internally.  When R is started up it grabs a piece of memory
and uses it to store the objects.  R performs its own memory management of
this piece of memory, growing and shrinking its size as needed.  Having
everything in memory is necessary because it is not really possible to
externally maintain all relevant &ldquo;environments&rdquo; of symbol/value pairs.
This difference also seems to make R <em>faster</em> than S.
</p>
<p>The down side is that if R crashes you will lose all the work for the
current session.  Saving and restoring the memory &ldquo;images&rdquo; (the functions
and data stored in R&rsquo;s internal memory at any time) can be a bit slow,
especially if they are big.  In S this does not happen, because everything
is saved in disk files and if you crash nothing is likely to happen to
them.  (In fact, one might conjecture that the S developers felt that the
price of changing their approach to persistent storage just to accommodate
lexical scope was far too expensive.)  Hence, when doing important work, you
might consider saving often (see <a href="#How-can-I-save-my-workspace_003f">How can I save my workspace?</a>) to
safeguard against possible crashes.  Other possibilities are logging your
sessions, or have your R commands stored in text files which can be read in
using <code>source()</code>.
</p>
<blockquote>
<p><b>Note:</b> If you run R from within Emacs (see <a href="#R-and-Emacs">R and Emacs</a>), you can save the
contents of the interaction buffer to a file and conveniently manipulate it
using <code>ess-transcript-mode</code>, as well as save source copies of all
functions and data used.
</p></blockquote>

<hr>
<a name="Models"></a>
<a name="Models-1"></a>
<h4 class="subsection">3.3.2 Models</h4>

<p>There are some differences in the modeling code, such as
</p>
<ul>
<li> Whereas in S, you would use <code>lm(y ~ x^3)</code> to regress <code>y</code> on
<code>x^3</code>, in R, you have to insulate powers of numeric vectors (using
<code>I()</code>), i.e., you have to use <code>lm(y ~ I(x^3))</code>.
</li><li> The glm family objects are implemented differently in R and S.  The same
functionality is available but the components have different names.
</li><li> Option <code>na.action</code> is set to <code>&quot;na.omit&quot;</code> by default in R, but not
set in S.
</li><li> Terms objects are stored differently.  In S a terms object is an expression
with attributes, in R it is a formula with attributes.  The attributes have
the same names but are mostly stored differently.
</li><li> Finally, in R <code>y ~ x + 0</code> is an alternative to <code>y ~ x - 1</code> for
specifying a model with no intercept.  Models with no parameters at all can
be specified by <code>y ~ 0</code>.
</li></ul>

<hr>
<a name="Others"></a>
<a name="Others-1"></a>
<h4 class="subsection">3.3.3 Others</h4>

<p>Apart from lexical scoping and its implications, R follows the S language
definition in the Blue and White Books as much as possible, and hence really
is an &ldquo;implementation&rdquo; of S.  There are some intentional differences where
the behavior of S is considered &ldquo;not clean&rdquo;.  In general, the rationale is
that R should help you detect programming errors, while at the same time
being as compatible as possible with S.
</p>
<p>Some known differences are the following.
</p>
<ul>
<li> In R, if <code>x</code> is a list, then <code>x[i] &lt;- NULL</code> and <code>x[[i]] &lt;-
NULL</code> remove the specified elements from <code>x</code>.  The first of these is
incompatible with S, where it is a no-op.  (Note that you can set elements
to <code>NULL</code> using <code>x[i] &lt;- list(NULL)</code>.)


</li><li> In S, the functions named <code>.First</code> and <code>.Last</code> in the <samp>.Data</samp>
directory can be used for customizing, as they are executed at the very
beginning and end of a session, respectively.

<p>In R, the startup mechanism is as follows.  Unless <samp>--no-environ</samp> was
given on the command line, R searches for site and user files to process for
setting environment variables.  Then, R searches for a site-wide startup
profile unless the command line option <samp>--no-site-file</samp> was given.
This code is loaded in package <strong>base</strong>.  Then, unless
<samp>--no-init-file</samp> was given, R searches for a user profile file, and
sources it into the user workspace.  It then loads a saved image of the user
workspace from <samp>.RData</samp> in case there is one (unless
<samp>--no-restore-data</samp> or <samp>--no-restore</samp> were specified).  Next,
a function <code>.First()</code> is run if found on the search path.  Finally,
function <code>.First.sys</code> in the <strong>base</strong> package is run.  When
terminating an R session, by default a function <code>.Last</code> is run if found
on the search path, followed by <code>.Last.sys</code>.  If needed, the functions
<code>.First()</code> and <code>.Last()</code> should be defined in the appropriate
startup profiles.  See the help pages for <code>.First</code> and <code>.Last</code> for
more details.
</p>
</li><li> In R, <code>T</code> and <code>F</code> are just variables being set to <code>TRUE</code> and
<code>FALSE</code>, respectively, but are not reserved words as in S and hence can
be overwritten by the user.  (This helps e.g. when you have factors with
levels <code>&quot;T&quot;</code> or <code>&quot;F&quot;</code>.)  Hence, when writing code you should
always use <code>TRUE</code> and <code>FALSE</code>.

</li><li> In R, <code>dyn.load()</code> can only load <em>shared objects</em>, as created for
example by <kbd>R CMD SHLIB</kbd>.

</li><li> In R, <code>attach()</code> currently only works for lists and data frames, but
not for directories.  (In fact, <code>attach()</code> also works for R data files
created with <code>save()</code>, which is analogous to attaching directories in
S.)  Also, you cannot attach at position 1.

</li><li> Categories do not exist in R, and never will as they are deprecated now in
S.  Use factors instead.

</li><li> In R, <code>For()</code> loops are not necessary and hence not supported.

</li><li> In R, <code>assign()</code> uses the argument <samp>envir=</samp> rather than
<samp>where=</samp> as in S.

</li><li> The random number generators are different, and the seeds have different
length.

</li><li> R passes integer objects to C as <code>int *</code> rather than <code>long *</code> as
in S.

</li><li> R has no single precision storage mode.  However, as of version 0.65.1,
there is a single precision interface to C/FORTRAN subroutines.

</li><li> By default, <code>ls()</code> returns the names of the objects in the current
(under R) and global (under S) environment, respectively.  For example,
given

<div class="example">
<pre class="example">x &lt;- 1; fun &lt;- function() {y &lt;- 1; ls()}
</pre></div>

<p>then <code>fun()</code> returns <code>&quot;y&quot;</code> in R and <code>&quot;x&quot;</code> (together with the
rest of the global environment) in S.
</p>
</li><li> R allows for zero-extent matrices (and arrays, i.e., some elements of the
<code>dim</code> attribute vector can be 0).  This has been determined a useful
feature as it helps reducing the need for special-case tests for empty
subsets.  For example, if <code>x</code> is a matrix, <code>x[, FALSE]</code> is not
<code>NULL</code> but a &ldquo;matrix&rdquo; with 0 columns.  Hence, such objects need to be
tested for by checking whether their <code>length()</code> is zero (which works in
both R and S), and not using <code>is.null()</code>.

</li><li> Named vectors are considered vectors in R but not in S (e.g.,
<code>is.vector(c(a = 1:3))</code> returns <code>FALSE</code> in S and <code>TRUE</code> in
R).

</li><li> Data frames are not considered as matrices in R (i.e., if <code>DF</code> is a
data frame, then <code>is.matrix(DF)</code> returns <code>FALSE</code> in R and
<code>TRUE</code> in S).

</li><li> R by default uses treatment contrasts in the unordered case, whereas S uses
the Helmert ones.  This is a deliberate difference reflecting the opinion
that treatment contrasts are more natural.

</li><li> In R, the argument of a replacement function which corresponds to the right
hand side must be named &lsquo;<samp>value</samp>&rsquo;.  E.g., <code>f(a) &lt;- b</code> is evaluated
as <code>a &lt;- &quot;f&lt;-&quot;(a, value = b)</code>.  S always takes the last argument,
irrespective of its name.

</li><li> In S, <code>substitute()</code> searches for names for substitution in the given
expression in three places: the actual and the default arguments of the
matching call, and the local frame (in that order).  R looks in the local
frame only, with the special rule to use a &ldquo;promise&rdquo; if a variable is not
evaluated.  Since the local frame is initialized with the actual arguments
or the default expressions, this is usually equivalent to S, until
assignment takes place.

</li><li> In S, the index variable in a <code>for()</code> loop is local to the inside of
the loop.  In R it is local to the environment where the <code>for()</code>
statement is executed.

</li><li> In S, <code>tapply(simplify=TRUE)</code> returns a vector where R returns a
one-dimensional array (which can have named dimnames).

</li><li> In S(-<small>PLUS</small>) the C locale is used, whereas in R the current operating
system locale is used for determining which characters are alphanumeric and
how they are sorted.  This affects the set of valid names for R objects (for
example accented chars may be allowed in R) and ordering in sorts and
comparisons (such as whether <code>&quot;aA&quot; &lt; &quot;Bb&quot;</code> is true or false).  From
version 1.2.0 the locale can be (re-)set in R by the <code>Sys.setlocale()</code>
function.

</li><li> In S, <code>missing(<var>arg</var>)</code> remains <code>TRUE</code> if <var>arg</var> is
subsequently modified; in R it doesn&rsquo;t.

</li><li> From R version 1.3.0, <code>data.frame</code> strips <code>I()</code> when creating
(column) names.

</li><li> In R, the string <code>&quot;NA&quot;</code> is not treated as a missing value in a
character variable.  Use <code>as.character(NA)</code> to create a missing
character value.

</li><li> R disallows repeated formal arguments in function calls.

</li><li> In S, <code>dump()</code>, <code>dput()</code> and <code>deparse()</code> are essentially
different interfaces to the same code.  In R from version 2.0.0, this is
only true if the same <code>control</code> argument is used, but by default it is
not.  By default <code>dump()</code> tries to write code that will evaluate to
reproduce the object, whereas <code>dput()</code> and <code>deparse()</code> default to
options for producing deparsed code that is readable.

</li><li> In R, indexing a vector, matrix, array or data frame with <code>[</code> using a
character vector index looks only for exact matches (whereas <code>[[</code> and
<code>$</code> allow partial matches).  In S, <code>[</code> allows partial matches.

</li><li> S has a two-argument version of <code>atan</code> and no <code>atan2</code>.  A call in
S such as <code>atan(x1, x2)</code> is equivalent to R&rsquo;s <code>atan2(x1, x2)</code>.
However, beware of named arguments since S&rsquo;s <code>atan(x = a, y = b)</code> is
equivalent to R&rsquo;s <code>atan2(y = a, x = b)</code> with the meanings of <code>x</code>
and <code>y</code> interchanged.  (R used to have undocumented support for a
two-argument <code>atan</code> with positional arguments, but this has been
withdrawn to avoid further confusion.)

</li><li> Numeric constants with no fractional and exponent (i.e., only integer)  part
are taken as integer in <small>S-PLUS</small> 6.x or later, but as double in R.

</li></ul>

<p>There are also differences which are not intentional, and result from
missing or incorrect code in R.  The developers would appreciate hearing
about any deficiencies you may find (in a written report fully documenting
the difference as you see it).  Of course, it would be useful if you were to
implement the change yourself and make sure it works.
</p>
<hr>
<a name="Is-there-anything-R-can-do-that-S_002dPLUS-cannot_003f"></a>
<a name="Is-there-anything-R-can-do-that-S_002dPLUS-cannot_003f-1"></a>
<h3 class="section">3.4 Is there anything R can do that <small>S-PLUS</small> cannot?</h3>

<p>Since almost anything you can do in R has source code that you could port to
<small>S-PLUS</small> with little effort there will never be much you can do in R that
you couldn&rsquo;t do in <small>S-PLUS</small> if you wanted to.  (Note that using lexical
scoping may simplify matters considerably, though.)
</p>
<p>R offers several graphics features that <small>S-PLUS</small> does not, such as finer
handling of line types, more convenient color handling (via palettes), gamma
correction for color, and, most importantly, mathematical annotation in plot
texts, via input expressions reminiscent of TeX constructs.  See the help
page for <code>plotmath</code>, which features an impressive on-line example.
More details can be found in Paul Murrell and Ross Ihaka (2000), &ldquo;An
Approach to Providing Mathematical Annotation in Plots&rdquo;,
<a href="http://www.amstat.org/publications/jcgs/"><em>Journal of
Computational and Graphical Statistics</em></a>, <strong>9</strong>, 582&ndash;599.
</p>
<hr>
<a name="What-is-R_002dplus_003f"></a>
<a name="What-is-R_002dplus_003f-1"></a>
<h3 class="section">3.5 What is R-plus?</h3>

<p>For a very long time, there was no such thing.
</p>
<p><a href="http://www.xlsolutions-corp.com/">XLSolutions Corporation</a> is currently
beta testing a commercially supported version of R named R+ (read R plus).
</p>
<p><a href="http://www.revolution-computing.com/">REvolution Computing</a> has
released <a href="http://www.revolution-computing.com/products/revolution-r.php">REvolution R</a>, an enterprise-class statistical analysis system based on R,
suitable for deployment in professional, commercial and regulated
environments.
</p>
<p><a href="http://www.random-technologies-llc.com/">Random Technologies</a> offers
<a href="http://random-technologies-llc.com/products/RStat/rstat">RStat</a>, an
enterprise-strength statistical computing environment which combines R with
enterprise-level validation, documentation, software support, and consulting
services, as well as related R-based products.
</p>
<p>See also
<a href="http://en.wikipedia.org/wiki/R_programming_language#Commercialized_versions_of_R">http://en.wikipedia.org/wiki/R_programming_language#Commercialized_versions_of_R</a>
for pointers to commercialized versions of R.
</p>
<hr>
<a name="R-Web-Interfaces"></a>
<a name="R-Web-Interfaces-1"></a>
<h2 class="chapter">4 R Web Interfaces</h2>

<p><strong>Rweb</strong> is developed and maintained by <a href="mailto:jeff@math.montana.edu">Jeff Banfield</a>.  The <a href="http://www.math.montana.edu/Rweb/">Rweb Home Page</a>
provides access to all three versions of Rweb&mdash;a simple text entry form
that returns output and graphs, a more sophisticated JavaScript version that
provides a multiple window environment, and a set of point and click modules
that are useful for introductory statistics courses and require no knowledge
of the R language.  All of the Rweb versions can analyze Web accessible
datasets if a <acronym>URL</acronym> is provided.
</p>
<p>The paper &ldquo;Rweb: Web-based Statistical Analysis&rdquo;, providing a detailed
explanation of the different versions of Rweb and an overview of how Rweb
works, was published in the Journal of Statistical Software
(<a href="http://www.jstatsoft.org/v04/i01/">http://www.jstatsoft.org/v04/i01/</a>).
</p>
<p><a href="mailto:ulfi@cs.tu-berlin.de">Ulf Bartel</a> has developed <strong>R-Online</strong>, a
simple on-line programming environment for R which intends to make the first
steps in statistical programming with R (especially with time series) as
easy as possible.  There is no need for a local installation since the only
requirement for the user is a JavaScript capable browser.  See
<a href="http://osvisions.com/r-online/">http://osvisions.com/r-online/</a> for more information.
</p>
<p><strong>Rcgi</strong> is a CGI WWW interface to R by <a href="mailto:mjr@dsl.pipex.com">MJ
Ray</a>.  It had the ability to use &ldquo;embedded code&rdquo;: you could mix user input
and code, allowing the <acronym>HTML</acronym> author to do anything from load in data sets
to enter most of the commands for users without writing CGI scripts.
Graphical output was possible in PostScript or GIF formats and the executed
code was presented to the user for revision.  However, it is not clear if
the project is still active.  Currently, a modified version of <strong>Rcgi</strong>
by <a href="mailto:mai@ms.uky.edu">Mai Zhou</a> (actually, two versions: one with
(bitmap) graphics and one without) as well as the original code are
available from <a href="http://www.ms.uky.edu/~statweb/">http://www.ms.uky.edu/~statweb/</a>.
</p>
<p>CGI-based web access to R is also provided at
<a href="http://hermes.sdu.dk/cgi-bin/go/">http://hermes.sdu.dk/cgi-bin/go/</a>.  There are many additional examples
of web interfaces to R which basically allow to submit R code to a remote
server, see for example the collection of links available from
<a href="http://biostat.mc.vanderbilt.edu/twiki/bin/view/Main/StatCompCourse">http://biostat.mc.vanderbilt.edu/twiki/bin/view/Main/StatCompCourse</a>.
</p>
<p><a href="http://www.warwick.ac.uk/go/dfirth">David Firth</a> has written
<a href="http://CRAN.R-project.org/package=CGIwithR"><strong>CGIwithR</strong></a>, an R add-on package available from <acronym>CRAN</acronym>.  It provides
some simple extensions to R to facilitate running R scripts through the CGI
interface to a web server, and allows submission of data using both GET and
POST methods.  It is easily installed using Apache under Linux and in
principle should run on any platform that supports R and a web server
provided that the installer has the necessary security permissions.  David&rsquo;s
paper &ldquo;CGIwithR: Facilities for Processing Web Forms Using R&rdquo; was
published in the Journal of Statistical Software
(<a href="http://www.jstatsoft.org/v08/i10/">http://www.jstatsoft.org/v08/i10/</a>).  The package is now maintained by
<a href="mailto:duncan@wald.ucdavis.edu">Duncan Temple Lang</a> and has a web page at
<a href="http://www.omegahat.org/CGIwithR/">http://www.omegahat.org/CGIwithR/</a>.
</p>
<p><a href="http://www.rpad.org/Rpad">Rpad</a>, developed and actively maintained by
Tom Short, provides a sophisticated environment which combines some of the
features of the previous approaches with quite a bit of JavaScript, allowing
for a <acronym>GUI</acronym>-like behavior (with sortable tables, clickable graphics,
editable output), etc.
</p>
<p>Jeff Horner is working on the R/Apache Integration Project which embeds the
R interpreter inside Apache 2 (and beyond).  A tutorial and presentation are
available from the project web page at
<a href="http://biostat.mc.vanderbilt.edu/twiki/bin/view/Main/RApacheProject">http://biostat.mc.vanderbilt.edu/twiki/bin/view/Main/RApacheProject</a>.
</p>
<p><a href="http://stats.math.uni-augsburg.de/Rserve/">Rserve</a> is a project
actively developed by Simon Urbanek.  It implements a TCP/IP server which
allows other programs to use facilities of R.  Clients are available from
the web site for Java and C++ (and could be written for other languages that
support TCP/IP sockets).
</p>
<p><a href="http://openstatserver.org/index.html">OpenStatServer</a> is being
developed by a team lead by Greg Warnes; it aims &ldquo;to provide clean access
to computational modules defined in a variety of computational environments
(R, SAS, Matlab, etc) via a single well-defined client interface&rdquo; and to
turn computational services into web services.
</p>
<p>Two projects use PHP to provide a web interface to R.
<a href="http://steve-chen.net/R_PHP/">R_PHP_Online</a> by Steve Chen (though it is
unclear if this project is still active) is somewhat similar to the above
Rcgi and Rweb.  <a href="http://dssm.unipa.it/R-php/?cmd=home">R-php</a> is
actively developed by Alfredo Pontillo and Angelo Mineo and provides both a
web interface to R and a set of pre-specified analyses that need no R code
input.
</p>
<p><a href="http://www.bioconductor.org/">webbioc</a> is &ldquo;an integrated web interface
for doing microarray analysis using several of the Bioconductor packages&rdquo;
and is designed to be installed at local sites as a shared computing
resource.
</p>
<p><a href="http://sysbio.mrc-bsu.cam.ac.uk/Rwui">Rwui</a> is a web application to
create user-friendly web interfaces for R scripts.  All code for the web
interface is created automatically.  There is no need for the user to do any
extra scripting or learn any new scripting techniques.
</p>
<p>The <a href="http://CRAN.R-project.org/package=R.rsp"><strong>R.rsp</strong></a> package by Henrik Bengtsson introduces &ldquo;R Server
Pages&rdquo;.  Analogous to Java Server Pages, an R server page is typically
<acronym>HTML</acronym> with embedded R code that gets evaluated when the page is
requested.  The package includes an internal cross-platform HTTP server
implemented in Tcl, so provides a good framework for including web-based
user interfaces in packages.  The approach is similar to the use of the
<a href="http://CRAN.R-project.org/package=brew"><strong>brew</strong></a> package with <a href="http://rapache.net/">Rapache</a> with the
advantage of cross-platform support and easy installation.
</p>
<p>The <a href="http://CRAN.R-project.org/package=Rook"><strong>Rook</strong></a> package by Jeffrey Horner provides a web server interface
borrowing heavily from Ruby&rsquo;s Rack project.
</p>
<p>Finally, <a href="http://code.google.com/p/concerto-platform/">Concerto</a> is a
user friendly open-source Web Interface to R developed at the Psychometrics
Centre of Cambridge University.  It was designed as an online platform to
design and run Computerized Adaptive Tests, but can be also used as a
general-purpose R Web Interface.  It allows R users with no programming or
web designing background to quickly develop flexible and powerful online
applications, websites, and psychometrics tests.  To maximize its
reliability, security, and performance, Concerto relies on the popular and
reliable open-source elements such as MySQL server (exchange and storage of
the data), <a href="http://rstudio.org/">Rstudio</a> (R code designing and
testing, file management), CKEditor (<acronym>HTML</acronym> Layer design), and PHP.
</p>
<p>See <a href="http://rwiki.sciviews.org/doku.php?id=faq-r#web_interfaces">http://rwiki.sciviews.org/doku.php?id=faq-r#web_interfaces</a> for
additional information.
</p>

<hr>
<a name="R-Add_002dOn-Packages"></a>
<a name="R-Add_002dOn-Packages-1"></a>
<h2 class="chapter">5 R Add-On Packages</h2>


<hr>
<a name="Which-add_002don-packages-exist-for-R_003f"></a>
<a name="Which-add_002don-packages-exist-for-R_003f-1"></a>
<h3 class="section">5.1 Which add-on packages exist for R?</h3>


<hr>
<a name="Add_002don-packages-in-R"></a>
<a name="Add_002don-packages-in-R-1"></a>
<h4 class="subsection">5.1.1 Add-on packages in R</h4>

<p>The R distribution comes with the following packages:
</p>
<dl compact="compact">
<dt><strong>base</strong></dt>
<dd><p>Base R functions (and datasets before R 2.0.0).
</p></dd>
<dt><strong>compiler</strong></dt>
<dd><p>R byte code compiler (added in R 2.13.0).
</p></dd>
<dt><strong>datasets</strong></dt>
<dd><p>Base R datasets (added in R 2.0.0).
</p></dd>
<dt><strong>grDevices</strong></dt>
<dd><p>Graphics devices for base and grid graphics (added in R 2.0.0).
</p></dd>
<dt><strong>graphics</strong></dt>
<dd><p>R functions for base graphics.
</p></dd>
<dt><strong>grid</strong></dt>
<dd><p>A rewrite of the graphics layout capabilities, plus some support for
interaction.
</p></dd>
<dt><strong>methods</strong></dt>
<dd><p>Formally defined methods and classes for R objects, plus other programming
tools, as described in the Green Book.
</p></dd>
<dt><strong>parallel</strong></dt>
<dd><p>Support for parallel computation, including by forking and by sockets, and
random-number generation (added in R 2.14.0).
</p></dd>
<dt><strong>splines</strong></dt>
<dd><p>Regression spline functions and classes.
</p></dd>
<dt><strong>stats</strong></dt>
<dd><p>R statistical functions.
</p></dd>
<dt><strong>stats4</strong></dt>
<dd><p>Statistical functions using S4 classes.
</p></dd>
<dt><strong>tcltk</strong></dt>
<dd><p>Interface and language bindings to Tcl/Tk <acronym>GUI</acronym> elements.
</p></dd>
<dt><strong>tools</strong></dt>
<dd><p>Tools for package development and administration.
</p></dd>
<dt><strong>utils</strong></dt>
<dd><p>R utility functions.
</p></dd>
</dl>
<p>These &ldquo;base packages&rdquo; were substantially reorganized in R 1.9.0.  The
former <strong>base</strong> was split into the four packages <strong>base</strong>,
<strong>graphics</strong>, <strong>stats</strong>, and <strong>utils</strong>.  Packages <strong>ctest</strong>,
<strong>eda</strong>, <strong>modreg</strong>, <strong>mva</strong>, <strong>nls</strong>, <strong>stepfun</strong> and <strong>ts</strong>
were merged into <strong>stats</strong>, package <strong>lqs</strong> returned to the recommended
package <a href="http://CRAN.R-project.org/package=MASS"><strong>MASS</strong></a>, and package <strong>mle</strong> moved to <strong>stats4</strong>.
</p>
<hr>
<a name="Add_002don-packages-from-CRAN"></a>
<a name="Add_002don-packages-from-CRAN-1"></a>
<h4 class="subsection">5.1.2 Add-on packages from <acronym>CRAN</acronym></h4>

<p>The <acronym>CRAN</acronym> <samp>src/contrib</samp> area contains a wealth of add-on packages,
including the following <em>recommended</em> packages which are to be included
in all binary distributions of R.
</p>
<dl compact="compact">
<dt><strong>KernSmooth</strong></dt>
<dd><p>Functions for kernel smoothing (and density estimation) corresponding to the
book &ldquo;Kernel Smoothing&rdquo; by M. P. Wand and M. C. Jones, 1995.
</p></dd>
<dt><strong>MASS</strong></dt>
<dd><p>Functions and datasets from the main package of Venables and Ripley,
&ldquo;Modern Applied Statistics with S&rdquo;.  (Contained in the <samp>VR</samp> bundle
for R versions prior to 2.10.0.)
</p></dd>
<dt><strong>Matrix</strong></dt>
<dd><p>A Matrix package.  (Recommended for R 2.9.0 or later.)
</p></dd>
<dt><strong>boot</strong></dt>
<dd><p>Functions and datasets for bootstrapping from the book &ldquo;Bootstrap Methods
and Their Applications&rdquo; by A. C. Davison and D. V. Hinkley, 1997, Cambridge
University Press.
</p></dd>
<dt><strong>class</strong></dt>
<dd><p>Functions for classification (<em>k</em>-nearest neighbor and LVQ).
(Contained in the <samp>VR</samp> bundle for R versions prior to 2.10.0.)
</p></dd>
<dt><strong>cluster</strong></dt>
<dd><p>Functions for cluster analysis.
</p></dd>
<dt><strong>codetools</strong></dt>
<dd><p>Code analysis tools.  (Recommended for R 2.5.0 or later.)
</p></dd>
<dt><strong>foreign</strong></dt>
<dd><p>Functions for reading and writing data stored by statistical software like
Minitab, S, SAS, SPSS, Stata, Systat, etc.
</p></dd>
<dt><strong>lattice</strong></dt>
<dd><p>Lattice graphics, an implementation of Trellis Graphics functions.
</p></dd>
<dt><strong>mgcv</strong></dt>
<dd><p>Routines for GAMs and other generalized ridge regression problems with
multiple smoothing parameter selection by GCV or UBRE.
</p></dd>
<dt><strong>nlme</strong></dt>
<dd><p>Fit and compare Gaussian linear and nonlinear mixed-effects models.
</p></dd>
<dt><strong>nnet</strong></dt>
<dd><p>Software for single hidden layer perceptrons (&ldquo;feed-forward neural
networks&rdquo;), and for multinomial log-linear models.  (Contained in the
<samp>VR</samp> bundle for R versions prior to 2.10.0.)
</p></dd>
<dt><strong>rpart</strong></dt>
<dd><p>Recursive PARTitioning and regression trees.
</p></dd>
<dt><strong>spatial</strong></dt>
<dd><p>Functions for kriging and point pattern analysis from &ldquo;Modern Applied
Statistics with S&rdquo; by W. Venables and B. Ripley.  (Contained in the
<samp>VR</samp> bundle for R versions prior to 2.10.0.)
</p></dd>
<dt><strong>survival</strong></dt>
<dd><p>Functions for survival analysis, including penalized likelihood.
</p></dd>
</dl>
<p>See the <a href="http://CRAN.R-project.org/web/packages/"><acronym>CRAN</acronym> contributed
packages page</a> for more information.
</p>
<p>Many of these packages are categorized into
<a href="http://CRAN.R-project.org/web/views/"><acronym>CRAN</acronym> Task Views</a>, allowing to
browse packages by topic and providing tools to automatically install all
packages for special areas of interest.
</p>
<p>Some <acronym>CRAN</acronym> packages that do not build out of the box on Windows, require
additional software, or are shipping third party libraries for Windows
cannot be made available on <acronym>CRAN</acronym> in form of a Windows binary packages.
Nevertheless, some of these packages are available at the &ldquo;<acronym>CRAN</acronym> extras&rdquo;
repository at <a href="http://www.stats.ox.ac.uk/pub/RWin/">http://www.stats.ox.ac.uk/pub/RWin/</a> kindly provided by
Brian D. Ripley.  Note that this repository is a default repository for
recent versions of R for Windows.
</p>
<hr>
<a name="Add_002don-packages-from-Omegahat"></a>
<a name="Add_002don-packages-from-Omegahat-1"></a>
<h4 class="subsection">5.1.3 Add-on packages from Omegahat</h4>

<p>The <a href="http://www.omegahat.org/">Omega Project for Statistical Computing</a>
provides a variety of open-source software for statistical applications,
with special emphasis on web-based software, Java, the Java virtual machine,
and distributed computing.  A <acronym>CRAN</acronym> style R package repository is
available via <a href="http://www.omegahat.org/R/">http://www.omegahat.org/R/</a>.  See
<a href="http://www.omegahat.org/">http://www.omegahat.org/</a> for information on most R packages available
from the Omega project.
</p>
<hr>
<a name="Add_002don-packages-from-Bioconductor"></a>
<a name="Add_002don-packages-from-Bioconductor-1"></a>
<h4 class="subsection">5.1.4 Add-on packages from Bioconductor</h4>

<p><a href="http://www.bioconductor.org/">Bioconductor</a> is an open source and open
development software project for the analysis and comprehension of genomic
data.  Most Bioconductor components are distributed as R add-on packages.
Initially most of the <a href="http://www.bioconductor.org/packages/bioc/">Bioconductor software packages</a> focused primarily on DNA microarray data
analysis.  As the project has matured, the functional scope of the software
packages broadened to include the analysis of all types of genomic data,
such as SAGE, sequence, or SNP data.  In addition, there are metadata
(annotation, CDF and probe) and experiment data packages.  See
<a href="http://www.bioconductor.org/download/">http://www.bioconductor.org/download/</a> for available packages and a
complete taxonomy via BioC Views.
</p>
<hr>
<a name="Other-add_002don-packages"></a>
<a name="Other-add_002don-packages-1"></a>
<h4 class="subsection">5.1.5 Other add-on packages</h4>

<p>Many more packages are available from places other than the three default
repositories discussed above (<acronym>CRAN</acronym>, Bioconductor and Omegahat).  In
particular, R-Forge provides a <acronym>CRAN</acronym> style repository at
<a href="http://R-Forge.R-project.org/">http://R-Forge.R-project.org/</a>.
</p>
<p>More code has been posted to the R-help mailing list, and can be obtained
from the mailing list archive.
</p>
<hr>
<a name="How-can-add_002don-packages-be-installed_003f"></a>
<a name="How-can-add_002don-packages-be-installed_003f-1"></a>
<h3 class="section">5.2 How can add-on packages be installed?</h3>

<p>(Unix-like only.)  The add-on packages on <acronym>CRAN</acronym> come as gzipped tar files
named <code><var>pkg</var>_<var>version</var>.tar.gz</code>, which may in fact be
&ldquo;bundles&rdquo; containing more than one package.  Let <var>path</var> be the path to
such a package file.  Provided that <code>tar</code> and <code>gzip</code> are
available on your system, type
</p>
<div class="example">
<pre class="example">$ R CMD INSTALL <var>path</var>/<var>pkg</var>_<var>version</var>.tar.gz
</pre></div>

<p>at the shell prompt to install to the library tree rooted at the first
directory in your library search path (see the help page for
<code>.libPaths()</code> for details on how the search path is determined).
</p>
<p>To install to another tree (e.g., your private one), use
</p>
<div class="example">
<pre class="example">$ R CMD INSTALL -l <var>lib</var> <var>path</var>/<var>pkg</var>_<var>version</var>.tar.gz
</pre></div>

<p>where <var>lib</var> gives the path to the library tree to install to.
</p>
<p>Even more conveniently, you can install and automatically update packages
from within R if you have access to repositories such as <acronym>CRAN</acronym>.  See the
help page for <code>available.packages()</code> for more information.
</p>





<hr>
<a name="How-can-add_002don-packages-be-used_003f"></a>
<a name="How-can-add_002don-packages-be-used_003f-1"></a>
<h3 class="section">5.3 How can add-on packages be used?</h3>

<p>To find out which additional packages are available on your system, type
</p>
<div class="example">
<pre class="example">library()
</pre></div>

<p>at the R prompt.  
</p>
<p>This produces something like
</p>
<blockquote>
<table class="cartouche" border="1"><tr><td>
<div class="smallexample">
<pre class="smallexample">Packages in `/home/me/lib/R':

mystuff       My own R functions, nicely packaged but not documented

Packages in `/usr/local/lib/R/library':

KernSmooth    Functions for kernel smoothing for Wand &amp; Jones (1995)
MASS          Main Package of Venables and Ripley's MASS
Matrix        Sparse and Dense Matrix Classes and Methods
base          The R Base package
boot          Bootstrap R (S-Plus) Functions (Canty)
class         Functions for Classification
cluster       Functions for clustering (by Rousseeuw et al.)
codetools     Code Analysis Tools for R
datasets      The R Datasets Package
foreign       Read Data Stored by Minitab, S, SAS, SPSS, Stata, Systat,
              dBase, ...
grDevices     The R Graphics Devices and Support for Colours and Fonts
graphics      The R Graphics Package
grid          The Grid Graphics Package
lattice       Lattice Graphics
methods       Formal Methods and Classes
mgcv          GAMs with GCV/AIC/REML smoothness estimation and GAMMs
              by PQL
nlme          Linear and Nonlinear Mixed Effects Models
nnet          Feed-forward Neural Networks and Multinomial Log-Linear
              Models
rpart         Recursive Partitioning
spatial       Functions for Kriging and Point Pattern Analysis
splines       Regression Spline Functions and Classes
stats         The R Stats Package
stats4        Statistical functions using S4 Classes
survival      Survival analysis, including penalised likelihood
tcltk         Tcl/Tk Interface
tools         Tools for Package Development
utils         The R Utils Package
</pre></div>
</td></tr></table>
</blockquote>

<p>You can &ldquo;load&rdquo; the installed package <var>pkg</var> by
</p>
<div class="example">
<pre class="example">library(<var>pkg</var>)
</pre></div>

<p>You can then find out which functions it provides by typing one of
</p>
<div class="example">
<pre class="example">library(help = <var>pkg</var>)
help(package = <var>pkg</var>)
</pre></div>

<p>You can unload the loaded package <var>pkg</var> by
</p>
<div class="example">
<pre class="example">detach(&quot;package:<var>pkg</var>&quot;, unload = TRUE)
</pre></div>

<p>(where <code>unload = TRUE</code> is needed only for packages with a namespace,
see <code>?unload</code>).
</p>
<hr>
<a name="How-can-add_002don-packages-be-removed_003f"></a>
<a name="How-can-add_002don-packages-be-removed_003f-1"></a>
<h3 class="section">5.4 How can add-on packages be removed?</h3>

<p>Use
</p>
<div class="example">
<pre class="example">$ R CMD REMOVE <var>pkg_1</var> &hellip; <var>pkg_n</var>
</pre></div>

<p>to remove the packages <var>pkg_1</var>, &hellip;, <var>pkg_n</var> from the library
tree rooted at the first directory given in <code>R_LIBS</code> if this is set and
non-null, and from the default library otherwise.  (Versions of R prior to
1.3.0 removed from the default library by default.)
</p>
<p>To remove from library <var>lib</var>, do
</p>
<div class="example">
<pre class="example">$ R CMD REMOVE -l <var>lib</var> <var>pkg_1</var> &hellip; <var>pkg_n</var>
</pre></div>

<hr>
<a name="How-can-I-create-an-R-package_003f"></a>
<a name="How-can-I-create-an-R-package_003f-1"></a>
<h3 class="section">5.5 How can I create an R package?</h3>

<p>A package consists of a subdirectory containing a file <samp>DESCRIPTION</samp>
and the subdirectories <samp>R</samp>, <samp>data</samp>, <samp>demo</samp>, <samp>exec</samp>,
<samp>inst</samp>, <samp>man</samp>, <samp>po</samp>, <samp>src</samp>, and <samp>tests</samp> (some of
which can be missing).  The package subdirectory may also contain files
<samp>INDEX</samp>, <samp>NAMESPACE</samp>, <samp>configure</samp>, <samp>cleanup</samp>,
<samp>LICENSE</samp>, <samp>LICENCE</samp>, <samp>COPYING</samp> and <samp>NEWS</samp>.
</p>
<p>See section &ldquo;Creating R packages&rdquo; in <cite>Writing R Extensions</cite>, for
details.  This manual is included in the R distribution, see <a href="#What-documentation-exists-for-R_003f">What documentation exists for R?</a>, and gives information on package structure,
the configure and cleanup mechanisms, and on automated package checking and
building.
</p>
<p>R version 1.3.0 has added the function <code>package.skeleton()</code> which will
set up directories, save data and code, and create skeleton help files for a
set of R functions and datasets.
</p>
<p>See <a href="#What-is-CRAN_003f">What is CRAN?</a>, for information on uploading a package to <acronym>CRAN</acronym>.
</p>
<hr>
<a name="How-can-I-contribute-to-R_003f"></a>
<a name="How-can-I-contribute-to-R_003f-1"></a>
<h3 class="section">5.6 How can I contribute to R?</h3>

<p>R is in active development and there is always a risk of bugs creeping in.
Also, the developers do not have access to all possible machines capable of
running R.  So, simply using it and communicating problems is certainly of
great value.
</p>
<p>The <a href="http://developer.R-project.org/">R Developer Page</a> acts as an
intermediate repository for more or less finalized ideas and plans for the R
statistical system.  It contains (pointers to) TODO lists, RFCs, various
other writeups, ideas lists, and SVN miscellanea.
</p>
<hr>
<a name="R-and-Emacs"></a>
<a name="R-and-Emacs-1"></a>
<h2 class="chapter">6 R and Emacs</h2>


<hr>
<a name="Is-there-Emacs-support-for-R_003f"></a>
<a name="Is-there-Emacs-support-for-R_003f-1"></a>
<h3 class="section">6.1 Is there Emacs support for R?</h3>

<p>There is an Emacs package called <acronym>ESS</acronym> (&ldquo;Emacs Speaks Statistics&rdquo;)
which provides a standard interface between statistical programs and
statistical processes.  It is intended to provide assistance for interactive
statistical programming and data analysis.  Languages supported include: S
dialects (R, S 3/4, and <small>S-PLUS</small> 3.x/4.x/5.x/6.x/7.x), LispStat dialects
(XLispStat, ViSta), SAS, Stata, and BUGS.
</p>
<p><acronym>ESS</acronym> grew out of the need for bug fixes and extensions to S-mode
4.8 (which was a <acronym>GNU</acronym> Emacs interface to S/<small>S-PLUS</small> version 3
only).  The current set of developers desired support for XEmacs, R, S4, and
MS Windows.  In addition, with new modes being developed for R, Stata, and
SAS, it was felt that a unifying interface and framework for the user
interface would benefit both the user and the developer, by helping both
groups conform to standard Emacs usage.  The end result is an increase in
efficiency for statistical programming and data analysis, over the usual
tools.
</p>
<p>R support contains code for editing R source code (syntactic indentation and
highlighting of source code, partial evaluations of code, loading and
error-checking of code, and source code revision maintenance) and
documentation (syntactic indentation and highlighting of source code,
sending examples to running <acronym>ESS</acronym> process, and previewing),
interacting with an inferior R process from within Emacs (command-line
editing, searchable command history, command-line completion of R object and
file names, quick access to object and search lists, transcript recording,
and an interface to the help system), and transcript manipulation (recording
and saving transcript files, manipulating and editing saved transcripts, and
re-evaluating commands from transcript files).
</p>
<p>The latest stable version of <acronym>ESS</acronym> is available via <acronym>CRAN</acronym> or the
<a href="http://ESS.R-project.org/">ESS web page</a>.  The <acronym>HTML</acronym> version of the
documentation can be found at <a href="http://stat.ethz.ch/ESS/">http://stat.ethz.ch/ESS/</a>.
</p>
<p><acronym>ESS</acronym> comes with detailed installation instructions.
</p>
<p>For help with <acronym>ESS</acronym>, send email to
<a href="mailto:ESS-help@stat.math.ethz.ch">ESS-help@stat.math.ethz.ch</a>.
</p>
<p>Please send bug reports and suggestions on <acronym>ESS</acronym> to
<a href="mailto:ESS-bugs@stat.math.ethz.ch">ESS-bugs@stat.math.ethz.ch</a>.  The easiest way to do this from is
within Emacs by typing <kbd>M-x ess-submit-bug-report</kbd> or using the [ESS] or
[iESS] pulldown menus.
</p>
<hr>
<a name="Should-I-run-R-from-within-Emacs_003f"></a>
<a name="Should-I-run-R-from-within-Emacs_003f-1"></a>
<h3 class="section">6.2 Should I run R from within Emacs?</h3>

<p>Yes, <em>definitely</em>.  Inferior R mode provides a readline/history
mechanism, object name completion, and syntax-based highlighting of the
interaction buffer using Font Lock mode, as well as a very convenient
interface to the R help system.
</p>
<p>Of course, it also integrates nicely with the mechanisms for editing R
source using Emacs.  One can write code in one Emacs buffer and send whole
or parts of it for execution to R; this is helpful for both data analysis
and programming.  One can also seamlessly integrate with a revision control
system, in order to maintain a log of changes in your programs and data, as
well as to allow for the retrieval of past versions of the code.
</p>
<p>In addition, it allows you to keep a record of your session, which can also
be used for error recovery through the use of the transcript mode.
</p>
<p>To specify command line arguments for the inferior R process, use <kbd>C-u
M-x R</kbd> for starting R.
</p>

<hr>
<a name="Debugging-R-from-within-Emacs"></a>
<a name="Debugging-R-from-within-Emacs-1"></a>
<h3 class="section">6.3 Debugging R from within Emacs</h3>

<p>To debug R &ldquo;from within Emacs&rdquo;, there are several possibilities.  To use
the Emacs GUD (Grand Unified Debugger) library with the recommended debugger
GDB, type <kbd>M-x gdb</kbd> and give the path to the R <em>binary</em> as
argument.  At the <code>gdb</code> prompt, set <code>R_HOME</code> and other
environment variables as needed (using e.g. <kbd>set env R_HOME
/path/to/R/</kbd>, but see also below), and start the binary with the desired
arguments (e.g., <kbd>run --quiet</kbd>).
</p>
<p>If you have <acronym>ESS</acronym>, you can do <kbd>C-u M-x R <span class="key">RET</span> - d <span class="key">SPC</span>
g d b <span class="key">RET</span></kbd> to start an inferior R process with arguments <samp>-d
gdb</samp>.
</p>
<p>A third option is to start an inferior R process via <acronym>ESS</acronym> (<kbd>M-x
R</kbd>) and then start GUD (<kbd>M-x gdb</kbd>) giving the R binary (using its full
path name) as the program to debug.  Use the program <code>ps</code> to find
the process number of the currently running R process then use the
<code>attach</code> command in gdb to attach it to that process.  One advantage of
this method is that you have separate <code>*R*</code> and <code>*gud-gdb*</code>
windows.  Within the <code>*R*</code> window you have all the <acronym>ESS</acronym>
facilities, such as object-name completion, that we know and love.
</p>
<p>When using GUD mode for debugging from within Emacs, you may find it most
convenient to use the directory with your code in it as the current working
directory and then make a symbolic link from that directory to the R
binary.  That way <samp>.gdbinit</samp> can stay in the directory with the code
and be used to set up the environment and the search paths for the source,
e.g. as follows:
</p>
<div class="example">
<pre class="example">set env R_HOME /opt/R
set env R_PAPERSIZE letter
set env R_PRINTCMD lpr
dir /opt/R/src/appl
dir /opt/R/src/main
dir /opt/R/src/nmath
dir /opt/R/src/unix
</pre></div>

<hr>
<a name="R-Miscellanea"></a>
<a name="R-Miscellanea-1"></a>
<h2 class="chapter">7 R Miscellanea</h2>








<hr>
<a name="How-can-I-set-components-of-a-list-to-NULL_003f"></a>
<a name="How-can-I-set-components-of-a-list-to-NULL_003f-1"></a>
<h3 class="section">7.1 How can I set components of a list to NULL?</h3>

<p>You can use
</p>
<div class="example">
<pre class="example">x[i] &lt;- list(NULL)
</pre></div>

<p>to set component <code>i</code> of the list <code>x</code> to <code>NULL</code>, similarly for
named components.  Do not set <code>x[i]</code> or <code>x[[i]]</code> to <code>NULL</code>,
because this will remove the corresponding component from the list.
</p>
<p>For dropping the row names of a matrix <code>x</code>, it may be easier to use
<code>rownames(x) &lt;- NULL</code>, similarly for column names.
</p>
<hr>
<a name="How-can-I-save-my-workspace_003f"></a>
<a name="How-can-I-save-my-workspace_003f-1"></a>
<h3 class="section">7.2 How can I save my workspace?</h3>

<p><code>save.image()</code> saves the objects in the user&rsquo;s <code>.GlobalEnv</code> to the
file <samp>.RData</samp> in the R startup directory.  (This is also what happens
after <kbd>q(&quot;yes&quot;)</kbd>.)  Using <code>save.image(<var>file</var>)</code> one can save the
image under a different name.
</p>
<hr>
<a name="How-can-I-clean-up-my-workspace_003f"></a>
<a name="How-can-I-clean-up-my-workspace_003f-1"></a>
<h3 class="section">7.3 How can I clean up my workspace?</h3>

<p>To remove all objects in the currently active environment (typically
<code>.GlobalEnv</code>), you can do
</p>
<div class="example">
<pre class="example">rm(list = ls(all = TRUE))
</pre></div>

<p>(Without <samp>all = TRUE</samp>, only the objects with names not starting with
a &lsquo;<samp>.</samp>&rsquo; are removed.)
</p>
<hr>
<a name="How-can-I-get-eval_0028_0029-and-D_0028_0029-to-work_003f"></a>
<a name="How-can-I-get-eval_0028_0029-and-D_0028_0029-to-work_003f-1"></a>
<h3 class="section">7.4 How can I get eval() and D() to work?</h3>

<p>Strange things will happen if you use <code>eval(print(x), envir = e)</code> or
<code>D(x^2, &quot;x&quot;)</code>.  The first one will either tell you that &quot;<code>x</code>&quot; is
not found, or print the value of the wrong <code>x</code>.  The other one will
likely return zero if <code>x</code> exists, and an error otherwise.
</p>
<p>This is because in both cases, the first argument is evaluated in the
calling environment first.  The result (which should be an object of mode
<code>&quot;expression&quot;</code> or <code>&quot;call&quot;</code>) is then evaluated or differentiated.
What you (most likely) really want is obtained by &ldquo;quoting&rdquo; the first
argument upon surrounding it with <code>expression()</code>.  For example,
</p>
<div class="example">
<pre class="example">R&gt; D(expression(x^2), &quot;x&quot;)
2 * x
</pre></div>

<p>Although this behavior may initially seem to be rather strange, it is
perfectly logical.  The &ldquo;intuitive&rdquo; behavior could easily be implemented,
but problems would arise whenever the expression is contained in a variable,
passed as a parameter, or is the result of a function call.  Consider for
instance the semantics in cases like
</p>
<div class="example">
<pre class="example">D2 &lt;- function(e, n) D(D(e, n), n)
</pre></div>

<p>or
</p>
<div class="example">
<pre class="example">g &lt;- function(y) eval(substitute(y), sys.frame(sys.parent(n = 2)))
g(a * b)
</pre></div>

<p>See the help page for <code>deriv()</code> for more examples.
</p>
<hr>
<a name="Why-do-my-matrices-lose-dimensions_003f"></a>
<a name="Why-do-my-matrices-lose-dimensions_003f-1"></a>
<h3 class="section">7.5 Why do my matrices lose dimensions?</h3>

<p>When a matrix with a single row or column is created by a subscripting
operation, e.g., <code>row &lt;- mat[2, ]</code>, it is by default turned into a
vector.  In a similar way if an array with dimension, say, 2&nbsp;x&nbsp;3&nbsp;x&nbsp;1&nbsp;x&nbsp;4<!-- /@w -->
is created by subscripting it will be coerced into a 2&nbsp;x&nbsp;3&nbsp;x&nbsp;4<!-- /@w --> array,
losing the unnecessary dimension.  After much discussion this has been
determined to be a <em>feature</em>.
</p>
<p>To prevent this happening, add the option <samp>drop = FALSE</samp> to the
subscripting.  For example,
</p>
<div class="example">
<pre class="example">rowmatrix &lt;- mat[2, , drop = FALSE]  # <span class="roman">creates a row matrix</span>
colmatrix &lt;- mat[, 2, drop = FALSE]  # <span class="roman">creates a column matrix</span>
a &lt;- b[1, 1, 1, drop = FALSE]        # <span class="roman">creates a 1 x 1 x 1 array</span>
</pre></div>

<p>The <samp>drop = FALSE</samp> option should be used defensively when
programming.  For example, the statement
</p>
<div class="example">
<pre class="example">somerows &lt;- mat[index, ]
</pre></div>

<p>will return a vector rather than a matrix if <code>index</code> happens to have
length 1, causing errors later in the code.  It should probably be rewritten
as
</p>
<div class="example">
<pre class="example">somerows &lt;- mat[index, , drop = FALSE]
</pre></div>

<hr>
<a name="How-does-autoloading-work_003f"></a>
<a name="How-does-autoloading-work_003f-1"></a>
<h3 class="section">7.6 How does autoloading work?</h3>

<p>R has a special environment called <code>.AutoloadEnv</code>.  Using
<kbd>autoload(<var>name</var>, <var>pkg</var>)</kbd>, where <var>name</var> and <var>pkg</var> are
strings giving the names of an object and the package containing it, stores
some information in this environment.  When R tries to evaluate <var>name</var>,
it loads the corresponding package <var>pkg</var> and reevaluates <var>name</var> in
the new package&rsquo;s environment.
</p>
<p>Using this mechanism makes R behave as if the package was loaded, but does
not occupy memory (yet).
</p>
<p>See the help page for <code>autoload()</code> for a very nice example.
</p>
<hr>
<a name="How-should-I-set-options_003f"></a>
<a name="How-should-I-set-options_003f-1"></a>
<h3 class="section">7.7 How should I set options?</h3>

<p>The function <code>options()</code> allows setting and examining a variety of
global &ldquo;options&rdquo; which affect the way in which R computes and displays its
results.  The variable <code>.Options</code> holds the current values of these
options, but should never directly be assigned to unless you want to drive
yourself crazy&mdash;simply pretend that it is a &ldquo;read-only&rdquo; variable.
</p>
<p>For example, given
</p>
<div class="example">
<pre class="example">test1 &lt;- function(x = pi, dig = 3) {
  oo &lt;- options(digits = dig); on.exit(options(oo));
  cat(.Options$digits, x, &quot;\n&quot;)
}
test2 &lt;- function(x = pi, dig = 3) {
  .Options$digits &lt;- dig
  cat(.Options$digits, x, &quot;\n&quot;)
}
</pre></div>

<p>we obtain:
</p>
<div class="example">
<pre class="example">R&gt; test1()
3 3.14 
R&gt; test2()
3 3.141593
</pre></div>

<p>What is really used is the <em>global</em> value of <code>.Options</code>, and using
<kbd>options(OPT = VAL)</kbd> correctly updates it.  Local copies of
<code>.Options</code>, either in <code>.GlobalEnv</code> or in a function environment
(frame), are just silently disregarded.
</p>
<hr>
<a name="How-do-file-names-work-in-Windows_003f"></a>
<a name="How-do-file-names-work-in-Windows_003f-1"></a>
<h3 class="section">7.8 How do file names work in Windows?</h3>

<p>As R uses C-style string handling, &lsquo;<samp>\</samp>&rsquo; is treated as an escape
character, so that for example one can enter a newline as &lsquo;<samp>\n</samp>&rsquo;.  When
you really need a &lsquo;<samp>\</samp>&rsquo;, you have to escape it with another &lsquo;<samp>\</samp>&rsquo;.
</p>
<p>Thus, in filenames use something like <code>&quot;c:\\data\\money.dat&quot;</code>.  You can
also replace &lsquo;<samp>\</samp>&rsquo; by &lsquo;<samp>/</samp>&rsquo; (<code>&quot;c:/data/money.dat&quot;</code>).
</p>
<hr>
<a name="Why-does-plotting-give-a-color-allocation-error_003f"></a>
<a name="Why-does-plotting-give-a-color-allocation-error_003f-1"></a>
<h3 class="section">7.9 Why does plotting give a color allocation error?</h3>

<p>On an X11 device, plotting sometimes, e.g., when running
<code>demo(&quot;image&quot;)</code>, results in &ldquo;Error: color allocation error&rdquo;.  This is
an X problem, and only indirectly related to R.  It occurs when applications
started prior to R have used all the available colors.  (How many colors are
available depends on the X configuration; sometimes only 256 colors can be
used.)
</p>
<p>One application which is notorious for &ldquo;eating&rdquo; colors is Netscape.  If
the problem occurs when Netscape is running, try (re)starting it with either
the <samp>-no-install</samp> (to use the default colormap) or the
<samp>-install</samp> (to install a private colormap) option.
</p>
<p>You could also set the <code>colortype</code> of <code>X11()</code> to
<code>&quot;pseudo.cube&quot;</code> rather than the default <code>&quot;pseudo&quot;</code>.  See the help
page for <code>X11()</code> for more information.
</p>



<hr>
<a name="How-do-I-convert-factors-to-numeric_003f"></a>
<a name="How-do-I-convert-factors-to-numeric_003f-1"></a>
<h3 class="section">7.10 How do I convert factors to numeric?</h3>

<p>It may happen that when reading numeric data into R (usually, when reading
in a file), they come in as factors.  If <code>f</code> is such a factor object,
you can use
</p>
<div class="example">
<pre class="example">as.numeric(as.character(f))
</pre></div>

<p>to get the numbers back.  More efficient, but harder to remember, is
</p>
<div class="example">
<pre class="example">as.numeric(levels(f))[as.integer(f)]
</pre></div>

<p>In any case, do not call <code>as.numeric()</code> or their likes directly for the
task at hand (as <code>as.numeric()</code> or <code>unclass()</code> give the internal
codes).
</p>
<hr>
<a name="Are-Trellis-displays-implemented-in-R_003f"></a>
<a name="Are-Trellis-displays-implemented-in-R_003f-1"></a>
<h3 class="section">7.11 Are Trellis displays implemented in R?</h3>

<p>The recommended package <a href="http://CRAN.R-project.org/package=lattice"><strong>lattice</strong></a> (which is based on base package
<strong>grid</strong>) provides graphical functionality that is compatible with most
Trellis commands.
</p>
<p>You could also look at <code>coplot()</code> and <code>dotchart()</code> which might do
at least some of what you want.  Note also that the R version of
<code>pairs()</code> is fairly general and provides most of the functionality of
<code>splom()</code>, and that R&rsquo;s default plot method has an argument <code>asp</code>
allowing to specify (and fix against device resizing) the aspect ratio of
the plot.
</p>
<p>(Because the word &ldquo;Trellis&rdquo; has been claimed as a trademark we do not use
it in R.  The name &ldquo;lattice&rdquo; has been chosen for the R equivalent.)
</p>
<hr>
<a name="What-are-the-enclosing-and-parent-environments_003f"></a>
<a name="What-are-the-enclosing-and-parent-environments_003f-1"></a>
<h3 class="section">7.12 What are the enclosing and parent environments?</h3>

<p>Inside a function you may want to access variables in two additional
environments: the one that the function was defined in (&ldquo;enclosing&rdquo;), and
the one it was invoked in (&ldquo;parent&rdquo;).
</p>
<p>If you create a function at the command line or load it in a package its
enclosing environment is the global workspace.  If you define a function
<code>f()</code> inside another function <code>g()</code> its enclosing environment is
the environment inside <code>g()</code>.  The enclosing environment for a function
is fixed when the function is created.  You can find out the enclosing
environment for a function <code>f()</code> using <code>environment(f)</code>.
</p>
<p>The &ldquo;parent&rdquo; environment, on the other hand, is defined when you invoke a
function.  If you invoke <code>lm()</code> at the command line its parent
environment is the global workspace, if you invoke it inside a function
<code>f()</code> then its parent environment is the environment inside
<code>f()</code>.  You can find out the parent environment for an invocation of a
function by using <code>parent.frame()</code> or <code>sys.frame(sys.parent())</code>.
</p>
<p>So for most user-visible functions the enclosing environment will be the
global workspace, since that is where most functions are defined.  The
parent environment will be wherever the function happens to be called from.
If a function <code>f()</code> is defined inside another function <code>g()</code> it
will probably be used inside <code>g()</code> as well, so its parent environment
and enclosing environment will probably be the same.
</p>
<p>Parent environments are important because things like model formulas need to
be evaluated in the environment the function was called from, since that&rsquo;s
where all the variables will be available.  This relies on the parent
environment being potentially different with each invocation.
</p>
<p>Enclosing environments are important because a function can use variables in
the enclosing environment to share information with other functions or with
other invocations of itself (see the section on lexical scoping).  This
relies on the enclosing environment being the same each time the function is
invoked.  (In C this would be done with static variables.)
</p>
<p>Scoping <em>is</em> hard.  Looking at examples helps.  It is particularly
instructive to look at examples that work differently in R and S and try to
see why they differ.  One way to describe the scoping differences between R
and S is to say that in S the enclosing environment is <em>always</em> the
global workspace, but in R the enclosing environment is wherever the
function was created.
</p>
<hr>
<a name="How-can-I-substitute-into-a-plot-label_003f"></a>
<a name="How-can-I-substitute-into-a-plot-label_003f-1"></a>
<h3 class="section">7.13 How can I substitute into a plot label?</h3>

<p>Often, it is desired to use the value of an R object in a plot label, e.g.,
a title.  This is easily accomplished using <code>paste()</code> if the label is a
simple character string, but not always obvious in case the label is an
expression (for refined mathematical annotation).  In such a case, either
use <code>parse()</code> on your pasted character string or use
<code>substitute()</code> on an expression.  For example, if <code>ahat</code> is an
estimator of your parameter <em>a</em> of interest, use
</p>
<div class="example">
<pre class="example">title(substitute(hat(a) == ahat, list(ahat = ahat)))
</pre></div>

<p>(note that it is &lsquo;<samp>==</samp>&rsquo; and not &lsquo;<samp>=</samp>&rsquo;).  Sometimes <code>bquote()</code>
gives a more compact form, e.g.,
</p>
<div class="example">
<pre class="example">title(bquote(hat(a) = .(ahat)))
</pre></div>

<p>where subexpressions enclosed in &lsquo;<samp>.()</samp>&rsquo; are replaced by their values.
</p>
<p>There are more examples in the mailing list archives.
</p>
<hr>
<a name="What-are-valid-names_003f"></a>
<a name="What-are-valid-names_003f-1"></a>
<h3 class="section">7.14 What are valid names?</h3>

<p>When creating data frames using <code>data.frame()</code> or <code>read.table()</code>,
R by default ensures that the variable names are syntactically valid.  (The
argument <samp>check.names</samp> to these functions controls whether variable
names are checked and adjusted by <code>make.names()</code> if needed.)
</p>
<p>To understand what names are &ldquo;valid&rdquo;, one needs to take into account that
the term &ldquo;name&rdquo; is used in several different (but related) ways in the
language:
</p>
<ol>
<li> A <em>syntactic name</em> is a string the parser interprets as this type of
expression.  It consists of letters, numbers, and the dot and (for versions
of R at least 1.9.0) underscore characters, and starts with either a letter
or a dot not followed by a number.  Reserved words are not syntactic names.
</li><li> An <em>object name</em> is a string associated with an object that is assigned
in an expression either by having the object name on the left of an
assignment operation or as an argument to the <code>assign()</code> function.  It
is usually a syntactic name as well, but can be any non-empty string if it
is quoted (and it is always quoted in the call to <code>assign()</code>).

</li><li> An <em>argument name</em> is what appears to the left of the equals sign when
supplying an argument in a function call (for example, <code>f(trim=.5)</code>).
Argument names are also usually syntactic names, but again can be anything
if they are quoted.

</li><li> An <em>element name</em> is a string that identifies a piece of an object (a
component of a list, for example.)  When it is used on the right of the
&lsquo;<samp>$</samp>&rsquo; operator, it must be a syntactic name, or quoted.  Otherwise,
element names can be any strings.  (When an object is used as a database, as
in a call to <code>eval()</code> or <code>attach()</code>, the element names become
object names.)

</li><li> Finally, a <em>file name</em> is a string identifying a file in the operating
system for reading, writing, etc.  It really has nothing much to do with
names in the language, but it is traditional to call these strings file
&ldquo;names&rdquo;.
</li></ol>

<hr>
<a name="Are-GAMs-implemented-in-R_003f"></a>
<a name="Are-GAMs-implemented-in-R_003f-1"></a>
<h3 class="section">7.15 Are GAMs implemented in R?</h3>

<p>Package <a href="http://CRAN.R-project.org/package=gam"><strong>gam</strong></a> from <acronym>CRAN</acronym> implements all the Generalized Additive
Models (GAM) functionality as described in the GAM chapter of the White
Book.  In particular, it implements backfitting with both local regression
and smoothing splines, and is extendable.  There is a <code>gam()</code> function
for GAMs in package <a href="http://CRAN.R-project.org/package=mgcv"><strong>mgcv</strong></a>, but it is not an exact clone of what is
described in the White Book (no <code>lo()</code> for example).  Package
<a href="http://CRAN.R-project.org/package=gss"><strong>gss</strong></a> can fit spline-based GAMs too.  And if you can accept
regression splines you can use <code>glm()</code>.  For Gaussian GAMs you can use
<code>bruto()</code> from package <a href="http://CRAN.R-project.org/package=mda"><strong>mda</strong></a>.
</p>
<hr>
<a name="Why-is-the-output-not-printed-when-I-source_0028_0029-a-file_003f"></a>
<a name="Why-is-the-output-not-printed-when-I-source_0028_0029-a-file_003f-1"></a>
<h3 class="section">7.16 Why is the output not printed when I source() a file?</h3>

<p>Most R commands do not generate any output. The command
</p>
<div class="example">
<pre class="example">1+1
</pre></div>

<p>computes the value 2 and returns it; the command
</p>
<div class="example">
<pre class="example">summary(glm(y~x+z, family=binomial))
</pre></div>

<p>fits a logistic regression model, computes some summary information and
returns an object of class <code>&quot;summary.glm&quot;</code> (see <a href="#How-should-I-write-summary-methods_003f">How should I write summary methods?</a>).
</p>
<p>If you type &lsquo;<samp>1+1</samp>&rsquo; or &lsquo;<samp>summary(glm(y~x+z, family=binomial))</samp>&rsquo; at the
command line the returned value is automatically printed (unless it is
<code>invisible()</code>), but in other circumstances, such as in a
<code>source()</code>d file or inside a function it isn&rsquo;t printed unless you
specifically print it.
</p>
<p>To print the value use
</p>
<div class="example">
<pre class="example">print(1+1)
</pre></div>

<p>or
</p>
<div class="example">
<pre class="example">print(summary(glm(y~x+z, family=binomial)))
</pre></div>

<p>instead, or use <code>source(<var>file</var>, echo=TRUE)</code>.
</p>
<hr>
<a name="Why-does-outer_0028_0029-behave-strangely-with-my-function_003f"></a>
<a name="Why-does-outer_0028_0029-behave-strangely-with-my-function_003f-1"></a>
<h3 class="section">7.17 Why does outer() behave strangely with my function?</h3>

<p>As the help for <code>outer()</code> indicates, it does not work on arbitrary
functions the way the <code>apply()</code> family does.  It requires functions
that are vectorized to work elementwise on arrays.  As you can see by
looking at the code, <code>outer(x, y, FUN)</code> creates two large vectors
containing every possible combination of elements of <code>x</code> and <code>y</code>
and then passes this to <code>FUN</code> all at once.  Your function probably
cannot handle two large vectors as parameters.
</p>
<p>If you have a function that cannot handle two vectors but can handle two
scalars, then you can still use <code>outer()</code> but you will need to wrap
your function up first, to simulate vectorized behavior.  Suppose your
function is
</p>
<div class="example">
<pre class="example">foo &lt;- function(x, y, happy) {
  stopifnot(length(x) == 1, length(y) == 1) # scalars only!
  (x + y) * happy
}
</pre></div>

<p>If you define the general function
</p>
<div class="example">
<pre class="example">wrapper &lt;- function(x, y, my.fun, ...) {
  sapply(seq_along(x), FUN = function(i) my.fun(x[i], y[i], ...))
}
</pre></div>

<p>then you can use <code>outer()</code> by writing, e.g.,
</p>
<div class="example">
<pre class="example">outer(1:4, 1:2, FUN = wrapper, my.fun = foo, happy = 10)
</pre></div>

<p>Scalar functions can also be vectorized using <code>Vectorize()</code>.
</p>
<hr>
<a name="Why-does-the-output-from-anova_0028_0029-depend-on-the-order-of-factors-in-the-model_003f"></a>
<a name="Why-does-the-output-from-anova_0028_0029-depend-on-the-order-of-factors-in-the-model_003f-1"></a>
<h3 class="section">7.18 Why does the output from anova() depend on the order of factors in the model?</h3>

<p>In a model such as <code>~A+B+A:B</code>, R will report the difference in sums of
squares between the models <code>~1</code>, <code>~A</code>, <code>~A+B</code> and
<code>~A+B+A:B</code>.  If the model were <code>~B+A+A:B</code>, R would report
differences between <code>~1</code>, <code>~B</code>, <code>~A+B</code>, and <code>~A+B+A:B</code>
. In the first case the sum of squares for <code>A</code> is comparing <code>~1</code>
and <code>~A</code>, in the second case it is comparing <code>~B</code> and
<code>~B+A</code>.  In a non-orthogonal design (i.e., most unbalanced designs)
these comparisons are (conceptually and numerically)  different.
</p>
<p>Some packages report instead the sums of squares based on comparing the full
model to the models with each factor removed one at a time (the famous &lsquo;Type
III sums of squares&rsquo; from SAS, for example).  These do not depend on the
order of factors in the model.  The question of which set of sums of squares
is the Right Thing provokes low-level holy wars on R-help from time to time.
</p>
<p>There is no need to be agitated about the particular sums of squares that R
reports.  You can compute your favorite sums of squares quite easily.  Any
two models can be compared with <code>anova(<var>model1</var>, <var>model2</var>)</code>,
and <code>drop1(<var>model1</var>)</code> will show the sums of squares resulting from
dropping single terms.
</p>
<hr>
<a name="How-do-I-produce-PNG-graphics-in-batch-mode_003f"></a>
<a name="How-do-I-produce-PNG-graphics-in-batch-mode_003f-1"></a>
<h3 class="section">7.19 How do I produce PNG graphics in batch mode?</h3>

<p>Under a Unix-like, if your installation supports the <code>type=&quot;cairo&quot;</code>
option to the <code>png()</code> device there should be no problems, and the
default settings should just work.  This option is not available for
versions of R prior to 2.7.0, or without support for cairo.  From R 2.7.0
<code>png()</code> by default uses the Quartz device on OS X, and that too works
in batch mode.
</p>
<p>Earlier versions of the <code>png()</code> device used the X11 driver, which is a
problem in batch mode or for remote operation.  If you have Ghostscript you
can use <code>bitmap()</code>, which produces a PostScript or PDF file then
converts it to any bitmap format supported by Ghostscript.  On some
installations this produces ugly output, on others it is perfectly
satisfactory.  Many systems now come with Xvfb from
<a href="http://www.x.org/Downloads.html">X.Org</a> (possibly as an optional
install), which is an X11 server that does not require a screen; and there
is the <a href="http://CRAN.R-project.org/package=GDD"><strong>GDD</strong></a> package from <acronym>CRAN</acronym>, which produces PNG, JPEG and GIF
bitmaps without X11.
</p>
<hr>
<a name="How-can-I-get-command-line-editing-to-work_003f"></a>
<a name="How-can-I-get-command-line-editing-to-work_003f-1"></a>
<h3 class="section">7.20 How can I get command line editing to work?</h3>

<p>The Unix-like command-line interface to R can only provide the inbuilt
command line editor which allows recall, editing and re-submission of prior
commands provided that the <acronym>GNU</acronym> readline library is available at
the time R is configured for compilation.  Note that the &lsquo;development&rsquo;
version of readline including the appropriate headers is needed: users of
Linux binary distributions will need to install packages such as
<code>libreadline-dev</code> (Debian) or <code>readline-devel</code> (Red Hat).
</p>
<hr>
<a name="How-can-I-turn-a-string-into-a-variable_003f"></a>
<a name="How-can-I-turn-a-string-into-a-variable_003f-1"></a>
<h3 class="section">7.21 How can I turn a string into a variable?</h3>

<p>If you have
</p>
<div class="example">
<pre class="example">varname &lt;- c(&quot;a&quot;, &quot;b&quot;, &quot;d&quot;)
</pre></div>

<p>you can do
</p>
<div class="example">
<pre class="example">get(varname[1]) + 2
</pre></div>

<p>for
</p>
<div class="example">
<pre class="example">a + 2
</pre></div>

<p>or
</p>
<div class="example">
<pre class="example">assign(varname[1], 2 + 2)
</pre></div>

<p>for
</p>
<div class="example">
<pre class="example">a &lt;- 2 + 2
</pre></div>

<p>or
</p>
<div class="example">
<pre class="example">eval(substitute(lm(y ~ x + variable),
                list(variable = as.name(varname[1]))))
</pre></div>

<p>for
</p>
<div class="example">
<pre class="example">lm(y ~ x + a)
</pre></div>

<p>At least in the first two cases it is often easier to just use a list, and
then you can easily index it by name
</p>
<div class="example">
<pre class="example">vars &lt;- list(a = 1:10, b = rnorm(100), d = LETTERS)
vars[[&quot;a&quot;]]
</pre></div>

<p>without any of this messing about.
</p>
<hr>
<a name="Why-do-lattice_002ftrellis-graphics-not-work_003f"></a>
<a name="Why-do-lattice_002ftrellis-graphics-not-work_003f-1"></a>
<h3 class="section">7.22 Why do lattice/trellis graphics not work?</h3>

<p>The most likely reason is that you forgot to tell R to display the graph.
Lattice functions such as <code>xyplot()</code> create a graph object, but do not
display it (the same is true of <a href="http://CRAN.R-project.org/package=ggplot2"><strong>ggplot2</strong></a> graphics, and Trellis
graphics in <small>S-PLUS</small>).  The <code>print()</code> method for the graph object
produces the actual display.  When you use these functions interactively at
the command line, the result is automatically printed, but in
<code>source()</code> or inside your own functions you will need an explicit
<code>print()</code> statement.
</p>
<hr>
<a name="How-can-I-sort-the-rows-of-a-data-frame_003f"></a>
<a name="How-can-I-sort-the-rows-of-a-data-frame_003f-1"></a>
<h3 class="section">7.23 How can I sort the rows of a data frame?</h3>

<p>To sort the rows within a data frame, with respect to the values in one or
more of the columns, simply use <code>order()</code> (e.g., <code>DF[order(DF$a,
DF[[&quot;b&quot;]]), ]</code> to sort the data frame <code>DF</code> on columns named <code>a</code>
and <code>b</code>).
</p>
<hr>
<a name="Why-does-the-help_002estart_0028_0029-search-engine-not-work_003f"></a>
<a name="Why-does-the-help_002estart_0028_0029-search-engine-not-work_003f-1"></a>
<h3 class="section">7.24 Why does the help.start() search engine not work?</h3>

<p>The browser-based search engine in <code>help.start()</code> utilizes a Java
applet.  In order for this to function properly, a compatible version of
Java must installed on your system and linked to your browser, and both Java
<em>and</em> JavaScript need to be enabled in your browser.
</p>
<p>There have been a number of compatibility issues with versions of Java and
of browsers.  
For further details please consult section &ldquo;Enabling search in HTML help&rdquo;
in <cite>R Installation and Administration</cite>.  This manual is included in the
R distribution, see <a href="#What-documentation-exists-for-R_003f">What documentation exists for R?</a>, and its
<acronym>HTML</acronym> version is linked from the <acronym>HTML</acronym> search page.
</p>
<hr>
<a name="Why-did-my-_002eRprofile-stop-working-when-I-updated-R_003f"></a>
<a name="Why-did-my-_002eRprofile-stop-working-when-I-updated-R_003f-1"></a>
<h3 class="section">7.25 Why did my .Rprofile stop working when I updated R?</h3>

<p>Did you read the <samp>NEWS</samp> file? For functions that are not in the
<strong>base</strong> package you need to specify the correct package namespace, since
the code will be run <em>before</em> the packages are loaded.  E.g.,
</p>
<div class="example">
<pre class="example">ps.options(horizontal = FALSE)
help.start()
</pre></div>

<p>needs to be
</p>
<div class="example">
<pre class="example">grDevices::ps.options(horizontal = FALSE)
utils::help.start()
</pre></div>

<p>(<code>graphics::ps.options(horizontal = FALSE)</code> in R 1.9.x).
</p>
<hr>
<a name="Where-have-all-the-methods-gone_003f"></a>
<a name="Where-have-all-the-methods-gone_003f-1"></a>
<h3 class="section">7.26 Where have all the methods gone?</h3>

<p>Many functions, particularly S3 methods, are now hidden in namespaces.  This
has the advantage that they cannot be called inadvertently with arguments of
the wrong class, but it makes them harder to view.
</p>
<p>To see the code for an S3 method (e.g., <code>[.terms</code>) use
</p>
<div class="example">
<pre class="example">getS3method(&quot;[&quot;, &quot;terms&quot;)
</pre></div>

<p>To see the code for an unexported function <code>foo()</code> in the namespace of
package <code>&quot;bar&quot;</code> use <code>bar:::foo</code>.  Don&rsquo;t use these constructions to
call unexported functions in your own code&mdash;they are probably unexported
for a reason and may change without warning.
</p>
<hr>
<a name="How-can-I-create-rotated-axis-labels_003f"></a>
<a name="How-can-I-create-rotated-axis-labels_003f-1"></a>
<h3 class="section">7.27 How can I create rotated axis labels?</h3>

<p>To rotate axis labels (using base graphics), you need to use <code>text()</code>,
rather than <code>mtext()</code>, as the latter does not support
<code>par(&quot;srt&quot;)</code>.
</p>
<div class="example">
<pre class="example">## <span class="roman">Increase bottom margin to make room for rotated labels</span>
par(mar = c(7, 4, 4, 2) + 0.1)
## <span class="roman">Create plot with no x axis and no x axis label</span>
plot(1 : 8, xaxt = &quot;n&quot;,  xlab = &quot;&quot;)
## <span class="roman">Set up x axis with tick marks alone</span>
axis(1, labels = FALSE)
## <span class="roman">Create some text labels</span>
labels &lt;- paste(&quot;Label&quot;, 1:8, sep = &quot; &quot;)
## <span class="roman">Plot x axis labels at default tick marks</span>
text(1:8, par(&quot;usr&quot;)[3] - 0.25, srt = 45, adj = 1,
     labels = labels, xpd = TRUE)
## <span class="roman">Plot x axis label at line 6 (of 7)</span>
mtext(1, text = &quot;X Axis Label&quot;, line = 6)
</pre></div>

<p>When plotting the x axis labels, we use <code>srt = 45</code> for text rotation
angle, <code>adj = 1</code> to place the right end of text at the tick marks, and
<code>xpd = TRUE</code> to allow for text outside the plot region.  You can adjust
the value of the <code>0.25</code> offset as required to move the axis labels up
or down relative to the x axis.  See <code>?par</code> for more information.
</p>
<p>Also see Figure 1 and associated code in Paul Murrell (2003), &ldquo;Integrating
grid Graphics Output with Base Graphics Output&rdquo;, <em>R News</em>,
<strong>3/2</strong>, 7&ndash;12.
</p>
<hr>
<a name="Why-is-read_002etable_0028_0029-so-inefficient_003f"></a>
<a name="Why-is-read_002etable_0028_0029-so-inefficient_003f-1"></a>
<h3 class="section">7.28 Why is read.table() so inefficient?</h3>

<p>By default, <code>read.table()</code> needs to read in everything as character
data, and then try to figure out which variables to convert to numerics or
factors.  For a large data set, this takes considerable amounts of time and
memory.  Performance can substantially be improved by using the
<code>colClasses</code> argument to specify the classes to be assumed for the
columns of the table.
</p>
<hr>
<a name="What-is-the-difference-between-package-and-library_003f"></a>
<a name="What-is-the-difference-between-package-and-library_003f-1"></a>
<h3 class="section">7.29 What is the difference between package and library?</h3>

<p>A <em>package</em> is a standardized collection of material extending R, e.g.
providing code, data, or documentation.  A <em>library</em> is a place
(directory) where R knows to find packages it can use (i.e., which were
<em>installed</em>).  R is told to use a package (to &ldquo;load&rdquo; it and add it to
the search path) via calls to the function <code>library</code>.  I.e.,
<code>library()</code> is employed to load a package from libraries containing
packages.
</p>
<p>See <a href="#R-Add_002dOn-Packages">R Add-On Packages</a>, for more details.  See also Uwe Ligges (2003), &ldquo;R
Help Desk: Package Management&rdquo;, <em>R News</em>, <strong>3/3</strong>, 37&ndash;39.
</p>
<hr>
<a name="I-installed-a-package-but-the-functions-are-not-there"></a>
<a name="I-installed-a-package-but-the-functions-are-not-there-1"></a>
<h3 class="section">7.30 I installed a package but the functions are not there</h3>

<p>To actually <em>use</em> the package, it needs to be <em>loaded</em> using
<code>library()</code>.
</p>
<p>See <a href="#R-Add_002dOn-Packages">R Add-On Packages</a> and <a href="#What-is-the-difference-between-package-and-library_003f">What is the difference between package and library?</a> for more information.
</p>
<hr>
<a name="Why-doesn_0027t-R-think-these-numbers-are-equal_003f"></a>
<a name="Why-doesn_0027t-R-think-these-numbers-are-equal_003f-1"></a>
<h3 class="section">7.31 Why doesn&rsquo;t R think these numbers are equal?</h3>

<p>The only numbers that can be represented exactly in R&rsquo;s numeric type are
integers and fractions whose denominator is a power of 2.  Other numbers
have to be rounded to (typically) 53 binary digits accuracy.  As a result,
two floating point numbers will not reliably be equal unless they have been
computed by the same algorithm, and not always even then.  For example
</p>
<div class="example">
<pre class="example">R&gt; a &lt;- sqrt(2)
R&gt; a * a == 2
[1] FALSE
R&gt; a * a - 2
[1] 4.440892e-16
</pre></div>

<p>The function <code>all.equal()</code> compares two objects using a numeric
tolerance of <code>.Machine$double.eps ^ 0.5</code>.  If you want much greater
accuracy than this you will need to consider error propagation carefully.
</p>
<p>For more information, see e.g. David Goldberg (1991), &ldquo;What Every
Computer Scientist Should Know About Floating-Point Arithmetic&rdquo;, <em>ACM
Computing Surveys</em>, <strong>23/1</strong>, 5&ndash;48, also available via
<a href="http://www.validlab.com/goldberg/paper.pdf">http://www.validlab.com/goldberg/paper.pdf</a>.
</p>
<p>To quote from &ldquo;The Elements of Programming Style&rdquo; by Kernighan and
Plauger:
</p>
<blockquote>
<p><em>10.0 times 0.1 is hardly ever 1.0</em>.
</p></blockquote>


<hr>
<a name="How-can-I-capture-or-ignore-errors-in-a-long-simulation_003f"></a>
<a name="How-can-I-capture-or-ignore-errors-in-a-long-simulation_003f-1"></a>
<h3 class="section">7.32 How can I capture or ignore errors in a long simulation?</h3>

<p>Use <code>try()</code>, which returns an object of class <code>&quot;try-error&quot;</code>
instead of an error, or preferably <code>tryCatch()</code>, where the return value
can be configured more flexibly.  For example
</p>
<div class="example">
<pre class="example">beta[i,] &lt;- tryCatch(coef(lm(formula, data)),
                     error = function(e) rep(NaN, 4))
</pre></div>

<p>would return the coefficients if the <code>lm()</code> call succeeded and would
return <code>c(NaN, NaN, NaN, NaN)</code> if it failed (presumably there are
supposed to be 4 coefficients in this example).
</p>
<hr>
<a name="Why-are-powers-of-negative-numbers-wrong_003f"></a>
<a name="Why-are-powers-of-negative-numbers-wrong_003f-1"></a>
<h3 class="section">7.33 Why are powers of negative numbers wrong?</h3>

<p>You are probably seeing something like
</p>
<div class="example">
<pre class="example">R&gt; -2^2
[1] -4
</pre></div>

<p>and misunderstanding the precedence rules for expressions in R.  Write
</p>
<div class="example">
<pre class="example">R&gt; (-2)^2
[1] 4
</pre></div>

<p>to get the square of <em>-2</em>.
</p>
<p>The precedence rules are documented in <code>?Syntax</code>, and to see how R
interprets an expression you can look at the parse tree
</p>
<div class="example">
<pre class="example">R&gt; as.list(quote(-2^2))
[[1]]
`-`

[[2]]
2^2
</pre></div>

<hr>
<a name="How-can-I-save-the-result-of-each-iteration-in-a-loop-into-a-separate-file_003f"></a>
<a name="How-can-I-save-the-result-of-each-iteration-in-a-loop-into-a-separate-file_003f-1"></a>
<h3 class="section">7.34 How can I save the result of each iteration in a loop into a separate file?</h3>

<p>One way is to use <code>paste()</code> (or <code>sprintf()</code>) to concatenate a stem
filename and the iteration number while <code>file.path()</code> constructs the
path.  For example, to save results into files <samp>result1.rda</samp>, &hellip;,
<samp>result100.rda</samp> in the subdirectory <samp>Results</samp> of the current
working directory, one can use
</p>
<div class="example">
<pre class="example">for(i in 1:100) {
  ## Calculations constructing &quot;some_object&quot; ...
  fp &lt;- file.path(&quot;Results&quot;, paste(&quot;result&quot;, i, &quot;.rda&quot;, sep = &quot;&quot;))
  save(list = &quot;some_object&quot;, file = fp)
}
</pre></div>

<hr>
<a name="Why-are-p_002dvalues-not-displayed-when-using-lmer_0028_0029_003f"></a>
<a name="Why-are-p_002dvalues-not-displayed-when-using-lmer_0028_0029_003f-1"></a>
<h3 class="section">7.35 Why are <em>p</em>-values not displayed when using lmer()?</h3>

<p>Doug Bates has kindly provided an extensive response in a post to the r-help
list, which can be reviewed at
<a href="https://stat.ethz.ch/pipermail/r-help/2006-May/094765.html">https://stat.ethz.ch/pipermail/r-help/2006-May/094765.html</a>.
</p>
<hr>
<a name="Why-are-there-unwanted-borders"></a>
<a name="Why-are-there-unwanted-borders_002c-lines-or-grid_002dlike-artifacts-when-viewing-a-plot-saved-to-a-PS-or-PDF-file_003f"></a>
<h3 class="section">7.36 Why are there unwanted borders, lines or grid-like artifacts when viewing a plot saved to a PS or PDF file?</h3>

<p>This can occur when using functions such as <code>polygon()</code>,
<code>filled.contour()</code>, <code>image()</code> or other functions which may call
these internally.  In the case of <code>polygon()</code>, you may observe unwanted
borders between the polygons even when setting the <code>border</code> argument to
<code>NA</code> or <code>&quot;transparent&quot;</code>.
</p>
<p>The source of the problem is the PS/PDF viewer when the plot is
anti-aliased.  The details for the solution will be different depending upon
the viewer used, the operating system and may change over time.  For some
common viewers, consider the following:
</p>
<dl compact="compact">
<dt>Acrobat Reader (cross platform)
<a name="index-Acrobat-Reader-_0028cross-platform_0029"></a>
</dt>
<dd><p>There are options in Preferences to enable/disable text smoothing, image
smoothing and line art smoothing.  Disable line art smoothing.
</p></dd>
<dt>Preview (OS X)
<a name="index-Preview-_0028OS-X_0029"></a>
</dt>
<dd><p>There is an option in Preferences to enable/disable anti-aliasing of text
and line art.  Disable this option. 
</p></dd>
<dt>GSview (cross platform)
<a name="index-GSview-_0028cross-platform_0029"></a>
</dt>
<dd><p>There are settings for Text Alpha and Graphics Alpha.  Change Graphics Alpha
from 4 bits to 1 bit to disable graphic anti-aliasing.
</p></dd>
<dt>gv (Unix-like X)
<a name="index-gv-_0028Unix_002dlike-X_0029"></a>
</dt>
<dd><p>There is an option to enable/disable anti-aliasing.  Disable this option.
</p></dd>
<dt>Evince (Linux/GNOME)
<a name="index-Evince-_0028Linux_002fGNOME_0029"></a>
</dt>
<dd><p>There is not an option to disable anti-aliasing in this viewer.
</p></dd>
<dt>Okular (Linux/KDE)
<a name="index-Okular-_0028Linux_002fKDE_0029"></a>
</dt>
<dd><p>There is not an option in the GUI to enable/disable anti-aliasing.  From a
console command line, use:
</p><div class="smallexample">
<pre class="smallexample">$ kwriteconfig --file okularpartrc --group 'Dlg Performance' \
               --key GraphicsAntialias Disabled
</pre></div>
<p>Then restart Okular.  Change the final word to &lsquo;<samp>Enabled</samp>&rsquo; to restore the
original setting.
</p></dd>
</dl>

<hr>
<a name="Why-does-backslash-behave-strangely-inside-strings_003f"></a>
<a name="Why-does-backslash-behave-strangely-inside-strings_003f-1"></a>
<h3 class="section">7.37 Why does backslash behave strangely inside strings?</h3>

<p>This question most often comes up in relation to file names (see <a href="#How-do-file-names-work-in-Windows_003f">How do file names work in Windows?</a>)  but it also happens that people complain that
they cannot seem to put a single &lsquo;<samp>\</samp>&rsquo; character into a text string
unless it happens to be followed by certain other characters.
</p>
<p>To understand this, you have to distinguish between character strings and
<em>representations</em> of character strings.  Mostly, the representation in
R is just the string with a single or double quote at either end, but there
are strings that cannot be represented that way, e.g., strings that
themselves contain the quote character.  So
</p>
<div class="example">
<pre class="example">&gt; str &lt;- &quot;This \&quot;text\&quot; is quoted&quot;
&gt; str
[1] &quot;This \&quot;text\&quot; is quoted&quot;
&gt; cat(str, &quot;\n&quot;)
This &quot;text&quot; is quoted
</pre></div>

<p>The <em>escape sequences</em> &lsquo;<samp>\&quot;</samp>&rsquo; and &lsquo;<samp>\n</samp>&rsquo; represent a double quote
and the newline character respectively. Printing text strings, using
<code>print()</code> or by typing the name at the prompt will use the escape
sequences too, but the <code>cat()</code> function will display the string
as-is. Notice that &lsquo;<samp>&quot;\n&quot;</samp>&rsquo; is a one-character string, not two; the
backslash is not actually in the string, it is just generated in the printed
representation.
</p>
<div class="example">
<pre class="example">&gt; nchar(&quot;\n&quot;)
[1] 1
&gt; substring(&quot;\n&quot;, 1, 1)
[1] &quot;\n&quot;
</pre></div>

<p>So how do you put a backslash in a string? For this, you have to escape the
escape character. I.e., you have to double the backslash.  as in
</p>
<div class="example">
<pre class="example">&gt; cat(&quot;\\n&quot;, &quot;\n&quot;)
\n
</pre></div>

<p>Some functions, particularly those involving regular expression matching,
themselves use metacharacters, which may need to be escaped by the backslash
mechanism.  In those cases you may need a <em>quadruple</em> backslash to
represent a single literal one.
</p>
<p>In versions of R up to 2.4.1 an unknown escape sequence like &lsquo;<samp>\p</samp>&rsquo; was
quietly interpreted as just &lsquo;<samp>p</samp>&rsquo;.  Current versions of R emit a warning.
</p>
<hr>
<a name="How-can-I-put-error-bars-or-confidence-bands-on-my-plot_003f"></a>
<a name="How-can-I-put-error-bars-or-confidence-bands-on-my-plot_003f-1"></a>
<h3 class="section">7.38 How can I put error bars or confidence bands on my plot?</h3>

<p>Some functions will display a particular kind of plot with error bars, such
as the <code>bar.err()</code> function in the <a href="http://CRAN.R-project.org/package=agricolae"><strong>agricolae</strong></a> package, the
<code>plotCI()</code> function in the <a href="http://CRAN.R-project.org/package=gplots"><strong>gplots</strong></a> package, the
<code>plotCI()</code> and <code>brkdn.plot()</code> functions in the <a href="http://CRAN.R-project.org/package=plotrix"><strong>plotrix</strong></a>
package and the <code>error.bars()</code>, <code>error.crosses()</code> and
<code>error.bars.by()</code> functions in the <a href="http://CRAN.R-project.org/package=psych"><strong>psych</strong></a> package.  Within
these types of functions, some will accept the measures of dispersion (e.g.,
<code>plotCI</code>), some will calculate the dispersion measures from the raw
values (<code>bar.err</code>, <code>brkdn.plot</code>), and some will do both
(<code>error.bars</code>).  Still other functions will just display error bars,
like the dispersion function in the <a href="http://CRAN.R-project.org/package=plotrix"><strong>plotrix</strong></a> package.  Most of the
above functions use the <code>arrows()</code> function in the base <strong>graphics</strong>
package to draw the error bars.
</p>
<p>The above functions all use the base graphics system.  The grid and lattice
graphics systems also have specific functions for displaying error bars,
e.g., the <code>grid.arrow()</code> function in the <strong>grid</strong> package, and the
<code>geom_errorbar()</code>, <code>geom_errorbarh()</code>, <code>geom_pointrange()</code>,
<code>geom_linerange()</code>, <code>geom_crossbar()</code> and <code>geom_ribbon()</code>
functions in the <a href="http://CRAN.R-project.org/package=ggplot2"><strong>ggplot2</strong></a> package.  In the lattice system, error
bars can be displayed with <code>Dotplot()</code> or <code>xYplot()</code> in the
<a href="http://CRAN.R-project.org/package=Hmisc"><strong>Hmisc</strong></a> package and <code>segplot()</code> in the <a href="http://CRAN.R-project.org/package=latticeExtra"><strong>latticeExtra</strong></a>
package.
</p>
<hr>
<a name="How-do-I-create-a-plot-with-two-y_002daxes_003f"></a>
<a name="How-do-I-create-a-plot-with-two-y_002daxes_003f-1"></a>
<h3 class="section">7.39 How do I create a plot with two y-axes?</h3>

<p>Creating a graph with two y-axes, i.e., with two sorts of data that are
scaled to the same vertical size and showing separate vertical axes on the
left and right sides of the plot that reflect the original scales of the
data, is possible in R but is not recommended.  The basic approach for
constructing such graphs is to use <code>par(new=TRUE)</code> (see <code>?par</code>);
functions <code>twoord.plot()</code> (in the <a href="http://CRAN.R-project.org/package=plotrix"><strong>plotrix</strong></a> package) and
<code>doubleYScale()</code> (in the <a href="http://CRAN.R-project.org/package=latticeExtra"><strong>latticeExtra</strong></a> package) automate the
process somewhat.  See
<a href="http://rwiki.sciviews.org/doku.php?id=tips:graphics-base:2yaxes">http://rwiki.sciviews.org/doku.php?id=tips:graphics-base:2yaxes</a> for
more information, including strong arguments against this sort of graph.
</p>
<hr>
<a name="How-do-I-access-the-source-code-for-a-function_003f"></a>
<a name="How-do-I-access-the-source-code-for-a-function_003f-1"></a>
<h3 class="section">7.40 How do I access the source code for a function?</h3>

<p>In most cases, typing the name of the function will print its source
code. However, code is sometimes hidden in a namespace, or compiled.  For a
complete overview on how to access source code, see Uwe Ligges (2006),
&ldquo;Help Desk: Accessing the sources&rdquo;, <em>R News</em>, <strong>6/4</strong>, 43&ndash;45
(<a href="http://CRAN.R-project.org/doc/Rnews/Rnews_2006-4.pdf">http://CRAN.R-project.org/doc/Rnews/Rnews_2006-4.pdf</a>).
</p>
<hr>
<a name="Why-does-summary_0028_0029-report-strange-results-for-the-R_005e2-estimate-when-I-fit-a-linear-model-with-no-intercept_003f"></a>
<a name="Why-does-summary_0028_0029-report-strange-results-for-the-R_005e2-estimate-when-I-fit-a-linear-model-with-no-intercept_003f-1"></a>
<h3 class="section">7.41 Why does summary() report strange results for the R^2 estimate when I fit a linear model with no intercept?</h3>

<p>As described in <code>?summary.lm</code>, when the intercept is zero (e.g., from
<code>y ~ x - 1</code> or <code>y ~ x + 0</code>), <code>summary.lm()</code> uses the formula
  R^2 = 1 - Sum(R[i]^2) / Sum((y[i])^2)
which is different from the usual
  R^2 = 1 - Sum(R[i]^2) / Sum((y[i] - mean(y))^2).
There are several reasons for this:
</p><ul>
<li> Otherwise the <em>R^2</em> could be negative (because the model with zero
intercept can fit <em>worse</em> than the constant-mean model it is implicitly
compared to).
</li><li> If you set the slope to zero in the model with a line through the origin you
get fitted values y*=0
</li><li> The model with constant, non-zero mean is not nested in the model with a
line through the origin.
</li></ul>

<p>All these come down to saying that if you know <em>a priori</em> that
<em>E[Y]=0</em> when <em>x=0</em> then the &lsquo;null&rsquo; model that you should compare
to the fitted line, the model where <em>x</em> doesn&rsquo;t explain any of the
variance, is the model where <em>E[Y]=0</em> everywhere.  (If you don&rsquo;t know a
priori that <em>E[Y]=0</em> when <em>x=0</em>, then you probably shouldn&rsquo;t be
fitting a line through the origin.)
</p>
<hr>
<a name="Why-is-R-apparently-not-releasing-memory_003f"></a>
<a name="Why-is-R-apparently-not-releasing-memory_003f-1"></a>
<h3 class="section">7.42 Why is R apparently not releasing memory?</h3>

<p>This question is often asked in different flavors along the lines of &ldquo;I
have removed objects in R and run <code>gc()</code> and yet <code>ps</code>/<code>top</code>
still shows the R process using a lot of memory&rdquo;, often on Linux machines.
</p>
<p>This is an artifact of the way the operating system (OS) allocates memory.
In general it is common that the OS is not capable of releasing all unused
memory.  In extreme cases it is possible that even if R frees almost all its
memory, the OS can not release any of it due to its design and thus tools
such as <code>ps</code> or <code>top</code> will report substantial amount of resident
RAM used by the R process even though R has released all that memory.  In
general such tools do <em>not</em> report the actual memory usage of the
process but rather what the OS is reserving for that process.
</p>
<p>The short answer is that this is a limitation of the memory allocator in the
operating system and there is nothing R can do about it. That space is
simply kept by the OS in the hope that R will ask for it later.  The
following paragraph gives more in-depth answer with technical details on how
this happens.
</p>
<p>Most systems use two separate ways to allocate memory. For allocation of
large chunks they will use <code>mmap</code> to map memory into the process
address space.  Such chunks can be released immediately when they are
completely free, because they can reside anywhere in the virtual memory.
However, this is a relatively expensive operation and many OSes have a limit
on the number of such allocated chunks, so this is only used for allocating
large memory regions.  For smaller allocations the system can expand the
data segment of the process (historically using the <code>brk</code> system call),
but this whole area is always contiguous.  The OS can only move the end of
this space, it cannot create any &ldquo;holes&rdquo;. Since this operation is fairly
cheap, it is used for allocations of small pieces of memory.  However, the
side-effect is that even if there is just one byte that is in use at the end
of the data segment, the OS cannot release any memory at all, because it
cannot change the address of that byte.  This is actually more common than
it may seem, because allocating a lot of intermediate objects, then
allocating a result object and removing all intermediate objects is a very
common practice.  Since the result is allocated at the end it will prevent
the OS from releasing any memory used by the intermediate objects.  In
practice, this is not necessarily a problem, because modern operating
systems can page out unused portions of the virtual memory so it does not
necessarily reduce the amount of real memory available for other
applications.  Typically, small objects such as strings or pairlists will be
affected by this behavior, whereas large objects such as long vectors will
be allocated using <code>mmap</code> and thus not affected.  On Linux (and
possibly other Unix-like systems) it is possible to use the <code>mallinfo</code>
system call (also see the <a href="http://rforge.net/mallinfo">mallinfo</a>
package) to query the allocator about the layout of the allocations,
including the actually used memory as well as unused memory that cannot be
released.
</p>
<hr>
<a name="R-Programming"></a>
<a name="R-Programming-1"></a>
<h2 class="chapter">8 R Programming</h2>


<hr>
<a name="How-should-I-write-summary-methods_003f"></a>
<a name="How-should-I-write-summary-methods_003f-1"></a>
<h3 class="section">8.1 How should I write summary methods?</h3>

<p>Suppose you want to provide a summary method for class <code>&quot;foo&quot;</code>.  Then
<code>summary.foo()</code> should not print anything, but return an object of
class <code>&quot;summary.foo&quot;</code>, <em>and</em> you should write a method
<code>print.summary.foo()</code> which nicely prints the summary information and
invisibly returns its object.  This approach is preferred over having
<code>summary.foo()</code> print summary information and return something useful,
as sometimes you need to grab something computed by <code>summary()</code> inside
a function or similar.  In such cases you don&rsquo;t want anything printed.
</p>
<hr>
<a name="How-can-I-debug-dynamically-loaded-code_003f"></a>
<a name="How-can-I-debug-dynamically-loaded-code_003f-1"></a>
<h3 class="section">8.2 How can I debug dynamically loaded code?</h3>

<p>Roughly speaking, you need to start R inside the debugger, load the code,
send an interrupt, and then set the required breakpoints.
</p>
<p>See section &ldquo;Finding entry points in dynamically loaded code&rdquo; in
<cite>Writing R Extensions</cite>.  This manual is included in the R distribution,
see <a href="#What-documentation-exists-for-R_003f">What documentation exists for R?</a>.
</p>
<hr>
<a name="How-can-I-inspect-R-objects-when-debugging_003f"></a>
<a name="How-can-I-inspect-R-objects-when-debugging_003f-1"></a>
<h3 class="section">8.3 How can I inspect R objects when debugging?</h3>

<p>The most convenient way is to call <code>R_PV</code> from the symbolic debugger.
</p>
<p>See section &ldquo;Inspecting R objects when debugging&rdquo; in <cite>Writing R
Extensions</cite>.
</p>
<hr>
<a name="How-can-I-change-compilation-flags_003f"></a>
<a name="How-can-I-change-compilation-flags_003f-1"></a>
<h3 class="section">8.4 How can I change compilation flags?</h3>

<p>Suppose you have C code file for dynloading into R, but you want to use
<code>R CMD SHLIB</code> with compilation flags other than the default ones (which
were determined when R was built).
</p>
<p>Starting with R 2.1.0, users can provide personal Makevars configuration
files in <samp>$<code>HOME</code>/.R</samp> to override the default flags.
See section &ldquo;Add-on packages&rdquo; in <cite>R Installation and Administration</cite>.
</p>
<p>For earlier versions of R, you could change the file
<samp><var>R_HOME</var>/etc/Makeconf</samp> to reflect your preferences, or (at least
for systems using <acronym>GNU</acronym> Make) override them by the environment
variable <code>MAKEFLAGS</code>.
See section &ldquo;Creating shared objects&rdquo; in <cite>Writing R Extensions</cite>.
</p>
<hr>
<a name="How-can-I-debug-S4-methods_003f"></a>
<a name="How-can-I-debug-S4-methods_003f-1"></a>
<h3 class="section">8.5 How can I debug S4 methods?</h3>

<p>Use the <code>trace()</code> function with argument <code>signature=</code> to add calls
to the browser or any other code to the method that will be dispatched for
the corresponding signature.  See <code>?trace</code> for details.
</p>
<hr>
<a name="R-Bugs"></a>
<a name="R-Bugs-1"></a>
<h2 class="chapter">9 R Bugs</h2>


<hr>
<a name="What-is-a-bug_003f"></a>
<a name="What-is-a-bug_003f-1"></a>
<h3 class="section">9.1 What is a bug?</h3>

<p>If R executes an illegal instruction, or dies with an operating system error
message that indicates a problem in the program (as opposed to something
like &ldquo;disk full&rdquo;), then it is certainly a bug.  If you call <code>.C()</code>,
<code>.Fortran()</code>, <code>.External()</code> or <code>.Call()</code> (or
<code>.Internal()</code>) yourself (or in a function you wrote), you can always
crash R by using wrong argument types (modes).  This is not a bug.
</p>
<p>Taking forever to complete a command can be a bug, but you must make certain
that it was really R&rsquo;s fault.  Some commands simply take a long time.  If
the input was such that you <em>know</em> it should have been processed
quickly, report a bug.  If you don&rsquo;t know whether the command should take a
long time, find out by looking in the manual or by asking for assistance.
</p>
<p>If a command you are familiar with causes an R error message in a case where
its usual definition ought to be reasonable, it is probably a bug.  If a
command does the wrong thing, that is a bug.  But be sure you know for
certain what it ought to have done.  If you aren&rsquo;t familiar with the
command, or don&rsquo;t know for certain how the command is supposed to work, then
it might actually be working right.  For example, people sometimes think
there is a bug in R&rsquo;s mathematics because they don&rsquo;t understand how
finite-precision arithmetic works.  Rather than jumping to conclusions, show
the problem to someone who knows for certain.  Unexpected results of
comparison of decimal numbers, for example <code>0.28 * 100 != 28</code> or
<code>0.1 + 0.2 != 0.3</code>, are not a bug.  See <a href="#Why-doesn_0027t-R-think-these-numbers-are-equal_003f">Why doesn't R think these numbers are equal?</a>, for more details.
</p>
<p>Finally, a command&rsquo;s intended definition may not be best for statistical
analysis.  This is a very important sort of problem, but it is also a matter
of judgment.  Also, it is easy to come to such a conclusion out of ignorance
of some of the existing features.  It is probably best not to complain about
such a problem until you have checked the documentation in the usual ways,
feel confident that you understand it, and know for certain that what you
want is not available.  If you are not sure what the command is supposed to
do after a careful reading of the manual this indicates a bug in the
manual.  The manual&rsquo;s job is to make everything clear.  It is just as
important to report documentation bugs as program bugs.  However, we know
that the introductory documentation is seriously inadequate, so you don&rsquo;t
need to report this.
</p>
<p>If the online argument list of a function disagrees with the manual, one of
them must be wrong, so report the bug.
</p>
<hr>
<a name="How-to-report-a-bug"></a>
<a name="How-to-report-a-bug-1"></a>
<h3 class="section">9.2 How to report a bug</h3>

<p>When you decide that there is a bug, it is important to report it and to
report it in a way which is useful.  What is most useful is an exact
description of what commands you type, starting with the shell command to
run R, until the problem happens.  Always include the version of R, machine,
and operating system that you are using; type <kbd>version</kbd> in R to print
this.
</p>
<p>The most important principle in reporting a bug is to report <em>facts</em>,
not hypotheses or categorizations.  It is always easier to report the facts,
but people seem to prefer to strain to posit explanations and report them
instead.  If the explanations are based on guesses about how R is
implemented, they will be useless; others will have to try to figure out
what the facts must have been to lead to such speculations.  Sometimes this
is impossible.  But in any case, it is unnecessary work for the ones trying
to fix the problem.
</p>
<p>For example, suppose that on a data set which you know to be quite large the
command
</p>
<div class="example">
<pre class="example">R&gt; data.frame(x, y, z, monday, tuesday)
</pre></div>

<p>never returns.  Do not report that <code>data.frame()</code> fails for large data
sets.  Perhaps it fails when a variable name is a day of the week.  If this
is so then when others got your report they would try out the
<code>data.frame()</code> command on a large data set, probably with no day of the
week variable name, and not see any problem.  There is no way in the world
that others could guess that they should try a day of the week variable
name.
</p>
<p>Or perhaps the command fails because the last command you used was a method
for <code>&quot;[&quot;()</code> that had a bug causing R&rsquo;s internal data structures to be
corrupted and making the <code>data.frame()</code> command fail from then on.
This is why others need to know what other commands you have typed (or read
from your startup file).
</p>
<p>It is very useful to try and find simple examples that produce apparently
the same bug, and somewhat useful to find simple examples that might be
expected to produce the bug but actually do not.  If you want to debug the
problem and find exactly what caused it, that is wonderful.  You should
still report the facts as well as any explanations or solutions.  Please
include an example that reproduces (e.g.,
<a href="http://en.wikipedia.org/wiki/Reproducibility">http://en.wikipedia.org/wiki/Reproducibility</a>) the problem, preferably
the simplest one you have found.
</p>
<p>Invoking R with the <samp>--vanilla</samp> option may help in isolating a bug.
This ensures that the site profile and saved data files are not read.
</p>
<p>Before you actually submit a bug report, you should check whether the bug
has already been reported and/or fixed.  First, try the &ldquo;Show open bugs
new-to-old&rdquo; or the search facility on <a href="http://bugs.R-project.org/">http://bugs.R-project.org/</a>.
Second, consult <a href="https://svn.R-project.org/R/trunk/doc/NEWS.Rd">https://svn.R-project.org/R/trunk/doc/NEWS.Rd</a>,
which records changes that will appear in the <em>next</em> release of R,
including bug fixes that do not appear on the Bug Tracker.  Third, if
possible try the current r-patched or r-devel version of R.  If a bug has
already been reported or fixed, please do not submit further bug reports on
it.  Finally, check carefully whether the bug is with R, or a contributed
package.  Bug reports on contributed packages should be sent first to the
package maintainer, and only submitted to the R-bugs repository by package
maintainers, mentioning the package in the subject line.
</p>
<p>A bug report can be generated using the function <code>bug.report()</code>.  For
reports on R this will open the Web page at
<a href="http://bugs.R-project.org/">http://bugs.R-project.org/</a>: for a contributed package it will open the
package&rsquo;s bug tracker Web page or help you compose an email to the
maintainer.
</p>
<p>There is a section of the bug repository for suggestions for enhancements
for R labelled &lsquo;<samp>wishlist</samp>&rsquo;.  Suggestions can be submitted in the same
ways as bugs, but please ensure that the subject line makes clear that this
is for the wishlist and not a bug report, for example by starting with
&lsquo;<samp>Wishlist:</samp>&rsquo;.
</p>
<p>Comments on and suggestions for the Windows port of R should be sent to
<a href="mailto:R-windows@R-project.org">R-windows@R-project.org</a>.
</p>
<p>Corrections to and comments on message translations should be sent to the
last translator (listed at the top of the appropriate &lsquo;<samp>.po</samp>&rsquo; file)  or
to the translation team as listed at
<a href="http://developer.R-project.org/TranslationTeams.html">http://developer.R-project.org/TranslationTeams.html</a>.
</p>
<hr>
<a name="Acknowledgments"></a>
<a name="Acknowledgments-1"></a>
<h2 class="chapter">10 Acknowledgments</h2>

<p>Of course, many many thanks to Robert and Ross for the R system, and to the
package writers and porters for adding to it.
</p>
<p>Special thanks go to Doug Bates, Peter Dalgaard, Paul Gilbert, Stefano
Iacus, Fritz Leisch, Jim Lindsey, Thomas Lumley, Martin Maechler, Brian
D. Ripley, Anthony Rossini, and Andreas Weingessel for their comments which
helped me improve this <acronym>FAQ</acronym>.
</p>
<p>More to come soon &hellip;
</p>
<hr>



</body>
</html>
